{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    " <li><a href=#pipeline> Understand the standard pipeline for calling variants in sequence data </a></li>\n",
    " <li><a href=#bwa>Use the BWA aligner to align whole genome sequence reads to a reference genome.</a></li>\n",
    " <li><a href=#samtools>use samtools and bcftools to call and filter variants from BWA alignment files  </a></li>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In tutorial 4, we learned how to use the [Burrows-Wheeler aligner](http://bio-bwa.sourceforge.net/) to map FASTQ reads to a reference genome. Now we will examine how the resulting alignment can serve as a starting point for identifying varyings in the genomic sequence data. We will follow the standard samtools workflow to identify variants in a yeast dataset: \n",
    "![pipeline](pipeline.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We begin the variant-calling pipeline with three files: \n",
    "\n",
    "* The yeast reference genome, stored in fasta format in the file **yeast.fasta**\n",
    "* First pair in yeast whole genome sequencing file, stored in fastq format, in the file **y1.fastq** \n",
    "* Second pair in yeast whole genome sequencing file, stored in fastq format, in the file **y2.fastq** "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Align the whole genome sequence reads to the yeast reference genome with BWA <a name='bwa'>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we must index the reference genome with the \"BwaIndexCommandline\" function. This function will generate the indices that the BWA aligner will need for rapid alignment of the WGS (whole genome sequence) reads. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio.Sequencing.Applications import BwaIndexCommandline\n",
    "reference_genome = \"yeast.fasta\"\n",
    "index_cmd = BwaIndexCommandline(infile=reference_genome, algorithm=\"bwtsw\")\n",
    "output=index_cmd()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#The following files were generated by the above indexing command: \n",
    "!ls yeast.fasta.*"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that the BWA index has been generated, we can run the BWA command to produce a paired end alignment of the WGS sample to the reference sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#First, we generate the .sai files for each WGS sample. \n",
    "from Bio.Sequencing.Applications import BwaAlignCommandline\n",
    "reference_genome = \"yeast.fasta\"\n",
    "read_file = \"y1.fastq\"\n",
    "output_sai_file = \"y1.sai\"\n",
    "align_cmd = BwaAlignCommandline(reference=reference_genome,read_file=read_file)\n",
    "print(align_cmd)\n",
    "output=align_cmd(stdout=output_sai_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Repeat for the paired reads in y2.fastq \n",
    "read_file=\"y2.fastq\"\n",
    "output_sai_file=\"y2.sai\"\n",
    "align_cmd = BwaAlignCommandline(reference=reference_genome,read_file=read_file)\n",
    "print(align_cmd)\n",
    "output=align_cmd(stdout=output_sai_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To call variants, we would like to convert our alignments from the binary .sai format to the .sam format that is compatible with samtools. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Import the BWA alignment tool from Biopython \n",
    "from Bio.Sequencing.Applications import BwaSampeCommandline\n",
    "reference_genome = \"yeast.fasta\"\n",
    "read_file1 = \"y1.fastq\"\n",
    "read_file2 = \"y2.fastq\"\n",
    "sai_file1 = \"y1.sai\"\n",
    "sai_file2 = \"y2.sai\"\n",
    "output_sam_file = \"output.sam\"\n",
    "sampe_cmd = BwaSampeCommandline(reference=reference_genome,\n",
    "                                 sai_file1=sai_file1, sai_file2=sai_file2,\n",
    "                                 read_file1=read_file1, read_file2=read_file2)\n",
    "print(sampe_cmd)\n",
    "output=sampe_cmd(stdout=output_sam_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The samtools file format is human readable, so we can examine the **output.sam** file that was generated by the command above. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "!head -n20 output.sam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can interpret the fields above as follows:\n",
    "![SAM alignment format](sam_alignment_format.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we sort the output.sam file into chromosome-position order, and convert to the binary .bam format for more efficient parsing. This can be accomplished with the **pysam** module. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pysam \n",
    "#sort the file \"output.sam\" and write the output to a binary file \"output.sorted.bam\" \n",
    "pysam.sort(\"-o\",\"output.sorted.bam\",\"output.sam\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we index the resulting **output.sorted.bam** file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "pysam.index(\"output.sorted.bam\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "As a sanity check, we can report some statistics about the alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#this command tells us how many reads in the fastq files did not align to the reference genome \n",
    "pysam.view(\"output.sorted.bam\",\"-c\",\"-f\",\"4\").strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#this command tells us how many reads in the fastq files aligned to the reference genome \n",
    "pysam.view(\"output.sorted.bam\",\"-c\",\"-F\",\"4\").strip()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This suggests that 92.6% of the reads in the FASTQ files align to the reference genome -- the data appears to be good quality!. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have aligned the samples to the reference genome, sorted the alignment file, converted it to binary format, and indexed it, we are ready to call variants with samtools. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Variant calling with Samtools  <a name='samtools'>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To convert your BAM file into genomic positions we first use mpileup to produce a BCF file that contains all of the locations in the genome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#make sure to include the catch_stdout flag to avoid printing a long output message that will slow down the Jupyter Noteobok. \n",
    "pysam.mpileup(\"output.sorted.bam\",\"-g\",\"-o\",\"output.bcf\",\"-f\",\"yeast.fasta\",catch_stdout=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use this information to call genotypes and reduce our list of sites to those found to be variant by passing this file into bcftools call. We pass the following flags to the bcftools call command: \n",
    "* -v (--variants-only) output variant sites only (as opposed to all sites in the genome) \n",
    "* -m (--multiallelic-caller) alternative model for mullti-allelic and rare variant calling\n",
    "* -O (--optimize) iteratively estimate the fraction of aberrant cells, down to the given fraction. Lowering this value from the default 1.0 to say, 0.3, can help discover more events but also increases noise\n",
    "* z zip the output "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
