{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Big Data for Biologists: Decoding Genomic Function- Class 11  \n",
    "\n",
    "## How are genes regulated? Transcription factors and regulatory elements\n",
    "\n",
    "##  Learning Objectives\n",
    "***Students should be able to***\n",
    "\n",
    "### How to represent a transcription factor's binding motif and scan sequences for matches \n",
    "\n",
    "<ol> \n",
    "    <li>  <a href=#TFMotif>Explain what transcription factors and transcription factor binding motifs are</a> </li>\n",
    " <li> <a  href=#PWM>Construct a position weight matrix (PWM) for a transcription factor by analyzing known transcription factor binding sequences </a></li>\n",
    " <li> <a href=#PSSM>Make a position-specific score matrices (PSSM) from a PWM to use for transcription factor motif-scanning  </a></li>\n",
    "    <li><a href=#Scan>Motif scanning along a DNA sequence </a></li>\n",
    " <li> <a href=#Biopython>Become familiar with Biopython functions and modules: Align, Motif</a> </li>\n",
    " <li> <a href=#ChIPseq>Understand what is a ChIP-seq experiment and how it can be used to identify transcription factor binding genome-wide</a> </li>\n",
    " <li> <a href=#ChipseqBrowser>Visualize transcription factor ChIP-seq experiment in the WashU Genome Browser</a> </li>\n",
    "</ol>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We have concluded our analysis of RNA-seq data. \n",
    "\n",
    "We will now transition to discussing the mehanisms that regulate gene expression.\n",
    "\n",
    "We will begin by examining transcription factors, proteins that control the rate of gene expression.  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "## We import our usual set of helper functions: \n",
    "import sys\n",
    "sys.path.append('../helpers')\n",
    "%matplotlib inline\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from viz_sequence import * "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What is a transcription factor?<a name='TFmotif' />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In molecular biology, <b>a transcription factor (TF) (or sequence-specific DNA-binding factor) is a protein that controls the rate of transcription of genetic information from DNA to messenger RNA, by binding to a specific DNA sequence</b>. Their function is to regulate - turn on and off - genes in order to make sure that they are expressed in the right cell at the right time and in the right amount throughout the life of the cell and the organism. Groups of TF's function in a coordinated fashion to direct cell division, cell growth, and cell death throughout life; cell migration and organization (body plan) during embryonic development; and intermittently in response to signals from outside the cell, such as a hormone. There are up to 2600 TFs in the human genome.\n",
    "\n",
    "TFs work alone or with other proteins in a complex, by promoting (as an <b>activator</b>), or blocking (as a <b>repressor</b>) the <b>recruitment of RNA polymerase</b> (the enzyme that performs the transcription of genetic information from DNA to RNA) to specific genes.\n",
    "\n",
    "A defining feature of TFs is that they contain at least one <b>DNA-binding domain (DBD), which has affinity to specific DNA short subsequences (~4-20 bp) collectively represented as a binding motif (pattern)</b>. TFs are grouped into classes based on their DBDs. Other proteins such as coactivators, chromatin remodelers, histone acetyltransferases, histone deacetylases, kinases, and methylases are also essential to gene regulation, but lack DNA-binding domains, and therefore are not TFs.\n",
    "\n",
    "TFs are of interest in medicine because TF mutations can cause specific diseases, and medications can be potentially targeted toward them.\n",
    "\n",
    "(Adapted from <a href=\"https://en.wikipedia.org/wiki/Transcription_factor\">https://en.wikipedia.org/wiki/Transcription_factor</a>)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## What is a transcription factor binding motif?<a name='TFMotif' />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![Transcription factor binds to DNA motif](../Images/10_tf_motif.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Each transcription factor (TF) has strong chemical binding affinity to some short DNA sequences and weaker affinity to others. Hence, the motif of a TF is not a single deterministic sequence. Instead, you should think of the binding motif of a TF as a collection of all possible subsequences (possibly of variable length), each of which has a different affinity score i.e. a distribution of affinity scores. This is clearly a clunky representation of a motif i.e. a very long list of subsequences with associated affinity scores. Can we develop a compact representation of a binding motif that faithfully summarizes the variable affinity of the TF to different subsequences.\n",
    "\n",
    "Let us assume we were able to conduct a sequencing experiment in some cell type, that provided us a sample of subsequences in the genome to which a TF binds.\n",
    "\n",
    "For example, we have found that the GATA transcription factor binds to the 4 sequences in the tf_binding_sequences.fa file.\n",
    "\n",
    "We could view this file using the !cat command. \n",
    "\n",
    "We can also view this file by reading it into Python and storing it as a list. Note in the code below that we are using the .strip() command to remove any whitespace in the file. We are then splitting the file at each of the line breaks so each line of the file is stored as a list in Python. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['>sequence1',\n",
       " 'TTGGATAA',\n",
       " '>sequence2',\n",
       " 'CGATAACCG',\n",
       " '>sequence3',\n",
       " 'CGATAT',\n",
       " '>sequence4',\n",
       " 'AAAGGATAT']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#tf_binding_sequences.fa is a file with known sequences that bind to the GATA transcription factor\n",
    "with open(\"tf_binding_sequences.fa\",'r') as infile:\n",
    "    sequences = infile.read().strip().split('\\n')\n",
    "sequences "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construct a transcription factor binding motif representation called a position weight matrix (PWM)<a name='PWM' />  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are given a collection of binding instances (subsequences) in the genome to which the TF binds and we want to derive a compact representation of the TF's binding motif.\n",
    "\n",
    "The main principles behind deriving a motif representation are as follows:\n",
    "1. The subsequences that a TF likes to bind to (high affinity) are typically quite similar with some mismatches, insertions and deletions (remember alignments?)\n",
    "2. The frequency of a subsequence in the collection of binding instances is proportional to the affinity of the TF to that subsequence i.e. If a TF has high affinity to a specific subsequence, that subsequence will be found more often in the collection of binding instances.\n",
    "\n",
    "Based on these two principles, deriving a compact representation of a motif involves:\n",
    "1. Using multiple sequence alignment to align all the binding instance subsequences (which could be of variable length) allowing for mismatches, insertions and deletions\n",
    "2. Counting frequencies of aligned subsequences, which can alternatively be recorded as counting frequencies of each type of nucleotide (A,C,G,T) at each position across all the aligned subsequences."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## From Multiple Sequence Alignment to PWM <a name='PWM' />\n",
    "\n",
    "We use the [MUSCLE algorithm](http://www.ebi.ac.uk/Tools/msa/muscle/) to perform multiple sequence alignment and generate a consensus sequence. Refer to Tutorial 5 to refresh your knowledge of multiple sequence alignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('',\n",
       " '\\nMUSCLE v3.8.31 by Robert C. Edgar\\n\\nhttp://www.drive5.com/muscle\\nThis software is donated to the public domain.\\nPlease cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.\\n\\ntf_binding_sequences 4 seqs, max length 9, avg  length 8\\n00:00:00     10 MB(1%)  Iter   1   10.00%  K-mer dist pass 1\\n00:00:00     10 MB(1%)  Iter   1  100.00%  K-mer dist pass 1\\n00:00:00     10 MB(1%)  Iter   1   10.00%  K-mer dist pass 2\\n00:00:00     10 MB(1%)  Iter   1  100.00%  K-mer dist pass 2\\n00:00:00     10 MB(1%)  Iter   1   33.33%  Align node       \\n00:00:00     11 MB(1%)  Iter   1   66.67%  Align node\\n00:00:00     11 MB(1%)  Iter   1  100.00%  Align node\\n00:00:00     11 MB(1%)  Iter   1  100.00%  Align node\\n00:00:00     11 MB(1%)  Iter   1   25.00%  Root alignment\\n00:00:00     11 MB(1%)  Iter   1   50.00%  Root alignment\\n00:00:00     11 MB(1%)  Iter   1   75.00%  Root alignment\\n00:00:00     11 MB(1%)  Iter   1  100.00%  Root alignment\\n00:00:00     11 MB(1%)  Iter   1  100.00%  Root alignment\\n00:00:00     11 MB(1%)  Iter   2  100.00%  Root alignment\\n00:00:00     11 MB(1%)  Iter   3   40.00%  Refine biparts\\n00:00:00     11 MB(1%)  Iter   3   60.00%  Refine biparts\\n00:00:00     11 MB(1%)  Iter   3   80.00%  Refine biparts\\n00:00:00     11 MB(1%)  Iter   3  100.00%  Refine biparts\\n00:00:00     11 MB(1%)  Iter   3  120.00%  Refine biparts\\n00:00:00     11 MB(1%)  Iter   3  100.00%  Refine biparts\\n')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We use the MUSCLE algorithm to perform multiple sequence alignment and generate a consensus sequence.\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "muscle_cline = MuscleCommandline(input=\"tf_binding_sequences.fa\",fastaout='tf_binding_sequences_out.fa')\n",
    "muscle_cline()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">sequence1\n",
      "-TTGGATAA---\n",
      ">sequence2\n",
      "---CGATAACCG\n",
      ">sequence3\n",
      "---CGATAT---\n",
      ">sequence4\n",
      "AAAGGATAT---\n"
     ]
    }
   ],
   "source": [
    "#Write a line of code to view the output of the alignment\n",
    "###BEGIN SOLUTION\n",
    "!cat tf_binding_sequences_out.fa\n",
    "###END SOLUTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We examine the aligned portions of the sequences using the AlignIO module:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aligned sequences:\n",
      "Alignment with 4 rows and 12 columns\n",
      "-TTGGATAA--- sequence1\n",
      "---CGATAACCG sequence2\n",
      "---CGATAT--- sequence3\n",
      "AAAGGATAT--- sequence4\n"
     ]
    }
   ],
   "source": [
    "#The BioPython module AlignIO is an interface for inputing and outputing sequence alignments\n",
    "#We use it here to view a formated alignment\n",
    "import io \n",
    "from Bio import AlignIO\n",
    "align = AlignIO.read(\"tf_binding_sequences_out.fa\",\"fasta\")\n",
    "print(\"aligned sequences:\")\n",
    "print(align)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The variable \"align\" represents an object in Python. We introduced objects in Class 4. As a reminder, an \"object\" in Python is a collection of data and methods that operate on that data. An object hass associated attributes and functions that we can examine by running the **help** command on the object.  You can think of a class in Python as a blueprint for a specific object. In Python terminology, an object is an \"instance\" of its class. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on MultipleSeqAlignment in module Bio.Align object:\n",
      "\n",
      "class MultipleSeqAlignment(builtins.object)\n",
      " |  MultipleSeqAlignment(records, alphabet=None, annotations=None, column_annotations=None)\n",
      " |  \n",
      " |  Represents a classical multiple sequence alignment (MSA).\n",
      " |  \n",
      " |  By this we mean a collection of sequences (usually shown as rows) which\n",
      " |  are all the same length (usually with gap characters for insertions or\n",
      " |  padding). The data can then be regarded as a matrix of letters, with well\n",
      " |  defined columns.\n",
      " |  \n",
      " |  You would typically create an MSA by loading an alignment file with the\n",
      " |  AlignIO module:\n",
      " |  \n",
      " |  >>> from Bio import AlignIO\n",
      " |  >>> align = AlignIO.read(\"Clustalw/opuntia.aln\", \"clustal\")\n",
      " |  >>> print(align)\n",
      " |  Alignment with 7 rows and 156 columns\n",
      " |  TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
      " |  TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
      " |  TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
      " |  TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
      " |  TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
      " |  TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
      " |  TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
      " |  \n",
      " |  In some respects you can treat these objects as lists of SeqRecord objects,\n",
      " |  each representing a row of the alignment. Iterating over an alignment gives\n",
      " |  the SeqRecord object for each row:\n",
      " |  \n",
      " |  >>> len(align)\n",
      " |  7\n",
      " |  >>> for record in align:\n",
      " |  ...     print(\"%s %i\" % (record.id, len(record)))\n",
      " |  ...\n",
      " |  gi|6273285|gb|AF191659.1|AF191 156\n",
      " |  gi|6273284|gb|AF191658.1|AF191 156\n",
      " |  gi|6273287|gb|AF191661.1|AF191 156\n",
      " |  gi|6273286|gb|AF191660.1|AF191 156\n",
      " |  gi|6273290|gb|AF191664.1|AF191 156\n",
      " |  gi|6273289|gb|AF191663.1|AF191 156\n",
      " |  gi|6273291|gb|AF191665.1|AF191 156\n",
      " |  \n",
      " |  You can also access individual rows as SeqRecord objects via their index:\n",
      " |  \n",
      " |  >>> print(align[0].id)\n",
      " |  gi|6273285|gb|AF191659.1|AF191\n",
      " |  >>> print(align[-1].id)\n",
      " |  gi|6273291|gb|AF191665.1|AF191\n",
      " |  \n",
      " |  And extract columns as strings:\n",
      " |  \n",
      " |  >>> print(align[:, 1])\n",
      " |  AAAAAAA\n",
      " |  \n",
      " |  Or, take just the first ten columns as a sub-alignment:\n",
      " |  \n",
      " |  >>> print(align[:, :10])\n",
      " |  Alignment with 7 rows and 10 columns\n",
      " |  TATACATTAA gi|6273285|gb|AF191659.1|AF191\n",
      " |  TATACATTAA gi|6273284|gb|AF191658.1|AF191\n",
      " |  TATACATTAA gi|6273287|gb|AF191661.1|AF191\n",
      " |  TATACATAAA gi|6273286|gb|AF191660.1|AF191\n",
      " |  TATACATTAA gi|6273290|gb|AF191664.1|AF191\n",
      " |  TATACATTAA gi|6273289|gb|AF191663.1|AF191\n",
      " |  TATACATTAA gi|6273291|gb|AF191665.1|AF191\n",
      " |  \n",
      " |  Combining this alignment slicing with alignment addition allows you to\n",
      " |  remove a section of the alignment. For example, taking just the first\n",
      " |  and last ten columns:\n",
      " |  \n",
      " |  >>> print(align[:, :10] + align[:, -10:])\n",
      " |  Alignment with 7 rows and 20 columns\n",
      " |  TATACATTAAGTGTACCAGA gi|6273285|gb|AF191659.1|AF191\n",
      " |  TATACATTAAGTGTACCAGA gi|6273284|gb|AF191658.1|AF191\n",
      " |  TATACATTAAGTGTACCAGA gi|6273287|gb|AF191661.1|AF191\n",
      " |  TATACATAAAGTGTACCAGA gi|6273286|gb|AF191660.1|AF191\n",
      " |  TATACATTAAGTGTACCAGA gi|6273290|gb|AF191664.1|AF191\n",
      " |  TATACATTAAGTATACCAGA gi|6273289|gb|AF191663.1|AF191\n",
      " |  TATACATTAAGTGTACCAGA gi|6273291|gb|AF191665.1|AF191\n",
      " |  \n",
      " |  Note - This object replaced the older Alignment object defined in module\n",
      " |  Bio.Align.Generic but is not fully backwards compatible with it.\n",
      " |  \n",
      " |  Note - This object does NOT attempt to model the kind of alignments used\n",
      " |  in next generation sequencing with multiple sequencing reads which are\n",
      " |  much shorter than the alignment, and where there is usually a consensus or\n",
      " |  reference sequence with special status.\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |      Combine two alignments with the same number of rows by adding them.\n",
      " |      \n",
      " |      If you have two multiple sequence alignments (MSAs), there are two ways to think\n",
      " |      about adding them - by row or by column. Using the extend method adds by row.\n",
      " |      Using the addition operator adds by column. For example,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> a1 = SeqRecord(Seq(\"AAAAC\"), id=\"Alpha\")\n",
      " |      >>> b1 = SeqRecord(Seq(\"AAA-C\"), id=\"Beta\")\n",
      " |      >>> c1 = SeqRecord(Seq(\"AAAAG\"), id=\"Gamma\")\n",
      " |      >>> a2 = SeqRecord(Seq(\"GT\"), id=\"Alpha\")\n",
      " |      >>> b2 = SeqRecord(Seq(\"GT\"), id=\"Beta\")\n",
      " |      >>> c2 = SeqRecord(Seq(\"GT\"), id=\"Gamma\")\n",
      " |      >>> left = MultipleSeqAlignment([a1, b1, c1],\n",
      " |      ...                             annotations={\"tool\": \"demo\", \"name\": \"start\"},\n",
      " |      ...                             column_annotations={\"stats\": \"CCCXC\"})\n",
      " |      >>> right = MultipleSeqAlignment([a2, b2, c2],\n",
      " |      ...                             annotations={\"tool\": \"demo\", \"name\": \"end\"},\n",
      " |      ...                             column_annotations={\"stats\": \"CC\"})\n",
      " |      \n",
      " |      Now, let's look at these two alignments:\n",
      " |      \n",
      " |      >>> print(left)\n",
      " |      Alignment with 3 rows and 5 columns\n",
      " |      AAAAC Alpha\n",
      " |      AAA-C Beta\n",
      " |      AAAAG Gamma\n",
      " |      >>> print(right)\n",
      " |      Alignment with 3 rows and 2 columns\n",
      " |      GT Alpha\n",
      " |      GT Beta\n",
      " |      GT Gamma\n",
      " |      \n",
      " |      And add them:\n",
      " |      \n",
      " |      >>> combined = left + right\n",
      " |      >>> print(combined)\n",
      " |      Alignment with 3 rows and 7 columns\n",
      " |      AAAACGT Alpha\n",
      " |      AAA-CGT Beta\n",
      " |      AAAAGGT Gamma\n",
      " |      \n",
      " |      For this to work, both alignments must have the same number of records (here\n",
      " |      they both have 3 rows):\n",
      " |      \n",
      " |      >>> len(left)\n",
      " |      3\n",
      " |      >>> len(right)\n",
      " |      3\n",
      " |      >>> len(combined)\n",
      " |      3\n",
      " |      \n",
      " |      The individual rows are SeqRecord objects, and these can be added together. Refer\n",
      " |      to the SeqRecord documentation for details of how the annotation is handled. This\n",
      " |      example is a special case in that both original alignments shared the same names,\n",
      " |      meaning when the rows are added they also get the same name.\n",
      " |      \n",
      " |      Any common annotations are preserved, but differing annotation is lost. This is\n",
      " |      the same behaviour used in the SeqRecord annotations and is designed to prevent\n",
      " |      accidental propagation of inappropriate values:\n",
      " |      \n",
      " |      >>> combined.annotations\n",
      " |      {'tool': 'demo'}\n",
      " |      \n",
      " |      Similarly any common per-column-annotations are combined:\n",
      " |      \n",
      " |      >>> combined.column_annotations\n",
      " |      {'stats': 'CCCXCCC'}\n",
      " |  \n",
      " |  __format__(self, format_spec)\n",
      " |      Return the alignment as a string in the specified file format.\n",
      " |      \n",
      " |      The format should be a lower case string supported as an output\n",
      " |      format by Bio.AlignIO (such as \"fasta\", \"clustal\", \"phylip\",\n",
      " |      \"stockholm\", etc), which is used to turn the alignment into a\n",
      " |      string.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> a = SeqRecord(Seq(\"ACTGCTAGCTAG\"), id=\"Alpha\", description=\"\")\n",
      " |      >>> b = SeqRecord(Seq(\"ACT-CTAGCTAG\"), id=\"Beta\", description=\"\")\n",
      " |      >>> c = SeqRecord(Seq(\"ACTGCTAGATAG\"), id=\"Gamma\", description=\"\")\n",
      " |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      " |      >>> print(format(align, \"fasta\"))\n",
      " |      >Alpha\n",
      " |      ACTGCTAGCTAG\n",
      " |      >Beta\n",
      " |      ACT-CTAGCTAG\n",
      " |      >Gamma\n",
      " |      ACTGCTAGATAG\n",
      " |      <BLANKLINE>\n",
      " |      >>> print(format(align, \"phylip\"))\n",
      " |       3 12\n",
      " |      Alpha      ACTGCTAGCT AG\n",
      " |      Beta       ACT-CTAGCT AG\n",
      " |      Gamma      ACTGCTAGAT AG\n",
      " |      <BLANKLINE>\n",
      " |  \n",
      " |  __getitem__(self, index)\n",
      " |      Access part of the alignment.\n",
      " |      \n",
      " |      Depending on the indices, you can get a SeqRecord object\n",
      " |      (representing a single row), a Seq object (for a single columns),\n",
      " |      a string (for a single characters) or another alignment\n",
      " |      (representing some part or all of the alignment).\n",
      " |      \n",
      " |      align[r,c] gives a single character as a string\n",
      " |      align[r] gives a row as a SeqRecord\n",
      " |      align[r,:] gives a row as a SeqRecord\n",
      " |      align[:,c] gives a column as a Seq\n",
      " |      \n",
      " |      align[:] and align[:,:] give a copy of the alignment\n",
      " |      \n",
      " |      Anything else gives a sub alignment, e.g.\n",
      " |      align[0:2] or align[0:2,:] uses only row 0 and 1\n",
      " |      align[:,1:3] uses only columns 1 and 2\n",
      " |      align[0:2,1:3] uses only rows 0 & 1 and only cols 1 & 2\n",
      " |      \n",
      " |      We'll use the following example alignment here for illustration:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> a = SeqRecord(Seq(\"AAAACGT\"), id=\"Alpha\")\n",
      " |      >>> b = SeqRecord(Seq(\"AAA-CGT\"), id=\"Beta\")\n",
      " |      >>> c = SeqRecord(Seq(\"AAAAGGT\"), id=\"Gamma\")\n",
      " |      >>> d = SeqRecord(Seq(\"AAAACGT\"), id=\"Delta\")\n",
      " |      >>> e = SeqRecord(Seq(\"AAA-GGT\"), id=\"Epsilon\")\n",
      " |      >>> align = MultipleSeqAlignment([a, b, c, d, e])\n",
      " |      \n",
      " |      You can access a row of the alignment as a SeqRecord using an integer\n",
      " |      index (think of the alignment as a list of SeqRecord objects here):\n",
      " |      \n",
      " |      >>> first_record = align[0]\n",
      " |      >>> print(\"%s %s\" % (first_record.id, first_record.seq))\n",
      " |      Alpha AAAACGT\n",
      " |      >>> last_record = align[-1]\n",
      " |      >>> print(\"%s %s\" % (last_record.id, last_record.seq))\n",
      " |      Epsilon AAA-GGT\n",
      " |      \n",
      " |      You can also access use python's slice notation to create a sub-alignment\n",
      " |      containing only some of the SeqRecord objects:\n",
      " |      \n",
      " |      >>> sub_alignment = align[2:5]\n",
      " |      >>> print(sub_alignment)\n",
      " |      Alignment with 3 rows and 7 columns\n",
      " |      AAAAGGT Gamma\n",
      " |      AAAACGT Delta\n",
      " |      AAA-GGT Epsilon\n",
      " |      \n",
      " |      This includes support for a step, i.e. align[start:end:step], which\n",
      " |      can be used to select every second sequence:\n",
      " |      \n",
      " |      >>> sub_alignment = align[::2]\n",
      " |      >>> print(sub_alignment)\n",
      " |      Alignment with 3 rows and 7 columns\n",
      " |      AAAACGT Alpha\n",
      " |      AAAAGGT Gamma\n",
      " |      AAA-GGT Epsilon\n",
      " |      \n",
      " |      Or to get a copy of the alignment with the rows in reverse order:\n",
      " |      \n",
      " |      >>> rev_alignment = align[::-1]\n",
      " |      >>> print(rev_alignment)\n",
      " |      Alignment with 5 rows and 7 columns\n",
      " |      AAA-GGT Epsilon\n",
      " |      AAAACGT Delta\n",
      " |      AAAAGGT Gamma\n",
      " |      AAA-CGT Beta\n",
      " |      AAAACGT Alpha\n",
      " |      \n",
      " |      You can also use two indices to specify both rows and columns. Using simple\n",
      " |      integers gives you the entry as a single character string. e.g.\n",
      " |      \n",
      " |      >>> align[3, 4]\n",
      " |      'C'\n",
      " |      \n",
      " |      This is equivalent to:\n",
      " |      \n",
      " |      >>> align[3][4]\n",
      " |      'C'\n",
      " |      \n",
      " |      or:\n",
      " |      \n",
      " |      >>> align[3].seq[4]\n",
      " |      'C'\n",
      " |      \n",
      " |      To get a single column (as a string) use this syntax:\n",
      " |      \n",
      " |      >>> align[:, 4]\n",
      " |      'CCGCG'\n",
      " |      \n",
      " |      Or, to get part of a column,\n",
      " |      \n",
      " |      >>> align[1:3, 4]\n",
      " |      'CG'\n",
      " |      \n",
      " |      However, in general you get a sub-alignment,\n",
      " |      \n",
      " |      >>> print(align[1:5, 3:6])\n",
      " |      Alignment with 4 rows and 3 columns\n",
      " |      -CG Beta\n",
      " |      AGG Gamma\n",
      " |      ACG Delta\n",
      " |      -GG Epsilon\n",
      " |      \n",
      " |      This should all seem familiar to anyone who has used the NumPy\n",
      " |      array or matrix objects.\n",
      " |  \n",
      " |  __init__(self, records, alphabet=None, annotations=None, column_annotations=None)\n",
      " |      Initialize a new MultipleSeqAlignment object.\n",
      " |      \n",
      " |      Arguments:\n",
      " |       - records - A list (or iterator) of SeqRecord objects, whose\n",
      " |                   sequences are all the same length.  This may be an be an\n",
      " |                   empty list.\n",
      " |       - alphabet - For backward compatibility only; its value should always\n",
      " |                    be None.\n",
      " |       - annotations - Information about the whole alignment (dictionary).\n",
      " |       - column_annotations - Per column annotation (restricted dictionary).\n",
      " |                    This holds Python sequences (lists, strings, tuples)\n",
      " |                    whose length matches the number of columns. A typical\n",
      " |                    use would be a secondary structure consensus string.\n",
      " |      \n",
      " |      You would normally load a MSA from a file using Bio.AlignIO, but you\n",
      " |      can do this from a list of SeqRecord objects too:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> a = SeqRecord(Seq(\"AAAACGT\"), id=\"Alpha\")\n",
      " |      >>> b = SeqRecord(Seq(\"AAA-CGT\"), id=\"Beta\")\n",
      " |      >>> c = SeqRecord(Seq(\"AAAAGGT\"), id=\"Gamma\")\n",
      " |      >>> align = MultipleSeqAlignment([a, b, c],\n",
      " |      ...                              annotations={\"tool\": \"demo\"},\n",
      " |      ...                              column_annotations={\"stats\": \"CCCXCCC\"})\n",
      " |      >>> print(align)\n",
      " |      Alignment with 3 rows and 7 columns\n",
      " |      AAAACGT Alpha\n",
      " |      AAA-CGT Beta\n",
      " |      AAAAGGT Gamma\n",
      " |      >>> align.annotations\n",
      " |      {'tool': 'demo'}\n",
      " |      >>> align.column_annotations\n",
      " |      {'stats': 'CCCXCCC'}\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |      Iterate over alignment rows as SeqRecord objects.\n",
      " |      \n",
      " |      e.g.\n",
      " |      \n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> a = SeqRecord(Seq(\"ACTGCTAGCTAG\"), id=\"Alpha\")\n",
      " |      >>> b = SeqRecord(Seq(\"ACT-CTAGCTAG\"), id=\"Beta\")\n",
      " |      >>> c = SeqRecord(Seq(\"ACTGCTAGATAG\"), id=\"Gamma\")\n",
      " |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      " |      >>> for record in align:\n",
      " |      ...    print(record.id)\n",
      " |      ...    print(record.seq)\n",
      " |      ...\n",
      " |      Alpha\n",
      " |      ACTGCTAGCTAG\n",
      " |      Beta\n",
      " |      ACT-CTAGCTAG\n",
      " |      Gamma\n",
      " |      ACTGCTAGATAG\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Return the number of sequences in the alignment.\n",
      " |      \n",
      " |      Use len(alignment) to get the number of sequences (i.e. the number of\n",
      " |      rows), and alignment.get_alignment_length() to get the length of the\n",
      " |      longest sequence (i.e. the number of columns).\n",
      " |      \n",
      " |      This is easy to remember if you think of the alignment as being like a\n",
      " |      list of SeqRecord objects.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return a representation of the object for debugging.\n",
      " |      \n",
      " |      The representation cannot be used with eval() to recreate the object,\n",
      " |      which is usually possible with simple python objects.  For example:\n",
      " |      \n",
      " |      <Bio.Align.MultipleSeqAlignment instance (2 records of length 14)\n",
      " |      at a3c184c>\n",
      " |      \n",
      " |      The hex string is the memory address of the object, see help(id).\n",
      " |      This provides a simple way to visually distinguish alignments of\n",
      " |      the same size.\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return a multi-line string summary of the alignment.\n",
      " |      \n",
      " |      This output is intended to be readable, but large alignments are\n",
      " |      shown truncated.  A maximum of 20 rows (sequences) and 50 columns\n",
      " |      are shown, with the record identifiers.  This should fit nicely on a\n",
      " |      single screen. e.g.\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> a = SeqRecord(Seq(\"ACTGCTAGCTAG\"), id=\"Alpha\")\n",
      " |      >>> b = SeqRecord(Seq(\"ACT-CTAGCTAG\"), id=\"Beta\")\n",
      " |      >>> c = SeqRecord(Seq(\"ACTGCTAGATAG\"), id=\"Gamma\")\n",
      " |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      " |      >>> print(align)\n",
      " |      Alignment with 3 rows and 12 columns\n",
      " |      ACTGCTAGCTAG Alpha\n",
      " |      ACT-CTAGCTAG Beta\n",
      " |      ACTGCTAGATAG Gamma\n",
      " |      \n",
      " |      See also the alignment's format method.\n",
      " |  \n",
      " |  append(self, record)\n",
      " |      Add one more SeqRecord object to the alignment as a new row.\n",
      " |      \n",
      " |      This must have the same length as the original alignment (unless this is\n",
      " |      the first record).\n",
      " |      \n",
      " |      >>> from Bio import AlignIO\n",
      " |      >>> align = AlignIO.read(\"Clustalw/opuntia.aln\", \"clustal\")\n",
      " |      >>> print(align)\n",
      " |      Alignment with 7 rows and 156 columns\n",
      " |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
      " |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
      " |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
      " |      TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
      " |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
      " |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
      " |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
      " |      >>> len(align)\n",
      " |      7\n",
      " |      \n",
      " |      We'll now construct a dummy record to append as an example:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> dummy = SeqRecord(Seq(\"N\"*156), id=\"dummy\")\n",
      " |      \n",
      " |      Now append this to the alignment,\n",
      " |      \n",
      " |      >>> align.append(dummy)\n",
      " |      >>> print(align)\n",
      " |      Alignment with 8 rows and 156 columns\n",
      " |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
      " |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
      " |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
      " |      TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
      " |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
      " |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
      " |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
      " |      NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN dummy\n",
      " |      >>> len(align)\n",
      " |      8\n",
      " |  \n",
      " |  extend(self, records)\n",
      " |      Add more SeqRecord objects to the alignment as rows.\n",
      " |      \n",
      " |      They must all have the same length as the original alignment. For\n",
      " |      example,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> a = SeqRecord(Seq(\"AAAACGT\"), id=\"Alpha\")\n",
      " |      >>> b = SeqRecord(Seq(\"AAA-CGT\"), id=\"Beta\")\n",
      " |      >>> c = SeqRecord(Seq(\"AAAAGGT\"), id=\"Gamma\")\n",
      " |      >>> d = SeqRecord(Seq(\"AAAACGT\"), id=\"Delta\")\n",
      " |      >>> e = SeqRecord(Seq(\"AAA-GGT\"), id=\"Epsilon\")\n",
      " |      \n",
      " |      First we create a small alignment (three rows):\n",
      " |      \n",
      " |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      " |      >>> print(align)\n",
      " |      Alignment with 3 rows and 7 columns\n",
      " |      AAAACGT Alpha\n",
      " |      AAA-CGT Beta\n",
      " |      AAAAGGT Gamma\n",
      " |      \n",
      " |      Now we can extend this alignment with another two rows:\n",
      " |      \n",
      " |      >>> align.extend([d, e])\n",
      " |      >>> print(align)\n",
      " |      Alignment with 5 rows and 7 columns\n",
      " |      AAAACGT Alpha\n",
      " |      AAA-CGT Beta\n",
      " |      AAAAGGT Gamma\n",
      " |      AAAACGT Delta\n",
      " |      AAA-GGT Epsilon\n",
      " |      \n",
      " |      Because the alignment object allows iteration over the rows as\n",
      " |      SeqRecords, you can use the extend method with a second alignment\n",
      " |      (provided its sequences have the same length as the original alignment).\n",
      " |  \n",
      " |  get_alignment_length(self)\n",
      " |      Return the maximum length of the alignment.\n",
      " |      \n",
      " |      All objects in the alignment should (hopefully) have the same\n",
      " |      length. This function will go through and find this length\n",
      " |      by finding the maximum length of sequences in the alignment.\n",
      " |      \n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> a = SeqRecord(Seq(\"ACTGCTAGCTAG\"), id=\"Alpha\")\n",
      " |      >>> b = SeqRecord(Seq(\"ACT-CTAGCTAG\"), id=\"Beta\")\n",
      " |      >>> c = SeqRecord(Seq(\"ACTGCTAGATAG\"), id=\"Gamma\")\n",
      " |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      " |      >>> align.get_alignment_length()\n",
      " |      12\n",
      " |      \n",
      " |      If you want to know the number of sequences in the alignment,\n",
      " |      use len(align) instead:\n",
      " |      \n",
      " |      >>> len(align)\n",
      " |      3\n",
      " |  \n",
      " |  sort(self, key=None, reverse=False)\n",
      " |      Sort the rows (SeqRecord objects) of the alignment in place.\n",
      " |      \n",
      " |      This sorts the rows alphabetically using the SeqRecord object id by\n",
      " |      default. The sorting can be controlled by supplying a key function\n",
      " |      which must map each SeqRecord to a sort value.\n",
      " |      \n",
      " |      This is useful if you want to add two alignments which use the same\n",
      " |      record identifiers, but in a different order. For example,\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> align1 = MultipleSeqAlignment([\n",
      " |      ...              SeqRecord(Seq(\"ACGT\"), id=\"Human\"),\n",
      " |      ...              SeqRecord(Seq(\"ACGG\"), id=\"Mouse\"),\n",
      " |      ...              SeqRecord(Seq(\"ACGC\"), id=\"Chicken\"),\n",
      " |      ...          ])\n",
      " |      >>> align2 = MultipleSeqAlignment([\n",
      " |      ...              SeqRecord(Seq(\"CGGT\"), id=\"Mouse\"),\n",
      " |      ...              SeqRecord(Seq(\"CGTT\"), id=\"Human\"),\n",
      " |      ...              SeqRecord(Seq(\"CGCT\"), id=\"Chicken\"),\n",
      " |      ...          ])\n",
      " |      \n",
      " |      If you simple try and add these without sorting, you get this:\n",
      " |      \n",
      " |      >>> print(align1 + align2)\n",
      " |      Alignment with 3 rows and 8 columns\n",
      " |      ACGTCGGT <unknown id>\n",
      " |      ACGGCGTT <unknown id>\n",
      " |      ACGCCGCT Chicken\n",
      " |      \n",
      " |      Consult the SeqRecord documentation which explains why you get a\n",
      " |      default value when annotation like the identifier doesn't match up.\n",
      " |      However, if we sort the alignments first, then add them we get the\n",
      " |      desired result:\n",
      " |      \n",
      " |      >>> align1.sort()\n",
      " |      >>> align2.sort()\n",
      " |      >>> print(align1 + align2)\n",
      " |      Alignment with 3 rows and 8 columns\n",
      " |      ACGCCGCT Chicken\n",
      " |      ACGTCGTT Human\n",
      " |      ACGGCGGT Mouse\n",
      " |      \n",
      " |      As an example using a different sort order, you could sort on the\n",
      " |      GC content of each sequence.\n",
      " |      \n",
      " |      >>> from Bio.SeqUtils import GC\n",
      " |      >>> print(align1)\n",
      " |      Alignment with 3 rows and 4 columns\n",
      " |      ACGC Chicken\n",
      " |      ACGT Human\n",
      " |      ACGG Mouse\n",
      " |      >>> align1.sort(key = lambda record: GC(record.seq))\n",
      " |      >>> print(align1)\n",
      " |      Alignment with 3 rows and 4 columns\n",
      " |      ACGT Human\n",
      " |      ACGC Chicken\n",
      " |      ACGG Mouse\n",
      " |      \n",
      " |      There is also a reverse argument, so if you wanted to sort by ID\n",
      " |      but backwards:\n",
      " |      \n",
      " |      >>> align1.sort(reverse=True)\n",
      " |      >>> print(align1)\n",
      " |      Alignment with 3 rows and 4 columns\n",
      " |      ACGG Mouse\n",
      " |      ACGT Human\n",
      " |      ACGC Chicken\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  column_annotations\n",
      " |      Dictionary of per-letter-annotation for the sequence.\n",
      " |  \n",
      " |  substitutions\n",
      " |      Return an Array with the number of substitutions of letters in the alignment.\n",
      " |      \n",
      " |      As an example, consider a multiple sequence alignment of three DNA sequences:\n",
      " |      \n",
      " |      >>> from Bio.Seq import Seq\n",
      " |      >>> from Bio.SeqRecord import SeqRecord\n",
      " |      >>> from Bio.Align import MultipleSeqAlignment\n",
      " |      >>> seq1 = SeqRecord(Seq(\"ACGT\"), id=\"seq1\")\n",
      " |      >>> seq2 = SeqRecord(Seq(\"A--A\"), id=\"seq2\")\n",
      " |      >>> seq3 = SeqRecord(Seq(\"ACGT\"), id=\"seq3\")\n",
      " |      >>> seq4 = SeqRecord(Seq(\"TTTC\"), id=\"seq4\")\n",
      " |      >>> alignment = MultipleSeqAlignment([seq1, seq2, seq3, seq4])\n",
      " |      >>> print(alignment)\n",
      " |      Alignment with 4 rows and 4 columns\n",
      " |      ACGT seq1\n",
      " |      A--A seq2\n",
      " |      ACGT seq3\n",
      " |      TTTC seq4\n",
      " |      \n",
      " |      >>> m = alignment.substitutions\n",
      " |      >>> print(m)\n",
      " |          A   C   G   T\n",
      " |      A 3.0 0.5 0.0 2.5\n",
      " |      C 0.5 1.0 0.0 2.0\n",
      " |      G 0.0 0.0 1.0 1.0\n",
      " |      T 2.5 2.0 1.0 1.0\n",
      " |      <BLANKLINE>\n",
      " |      \n",
      " |      Note that the matrix is symmetric, with counts divided equally on both\n",
      " |      sides of the diagonal. For example, the total number of substitutions\n",
      " |      between A and T in the alignment is 3.5 + 3.5 = 7.\n",
      " |      \n",
      " |      Any weights associated with the sequences are taken into account when\n",
      " |      calculating the substitution matrix.  For example, given the following\n",
      " |      multiple sequence alignment::\n",
      " |      \n",
      " |          GTATC  0.5\n",
      " |          AT--C  0.8\n",
      " |          CTGTC  1.0\n",
      " |      \n",
      " |      For the first column we have::\n",
      " |      \n",
      " |          ('A', 'G') : 0.5 * 0.8 = 0.4\n",
      " |          ('C', 'G') : 0.5 * 1.0 = 0.5\n",
      " |          ('A', 'C') : 0.8 * 1.0 = 0.8\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(align)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **help** command gives us a detailed description of the attributes and functions associated iwth the align object. If you want to visualize the attributes in a more compact manner, you can use call the \"__dict__\" command on the align object to visualize a dictionary representation of this object, with the attributes as keys and their values as the dictionary values. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Bio.Align.MultipleSeqAlignment'>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'_records': [SeqRecord(seq=Seq('-TTGGATAA---'), id='sequence1', name='sequence1', description='sequence1', dbxrefs=[]),\n",
       "  SeqRecord(seq=Seq('---CGATAACCG'), id='sequence2', name='sequence2', description='sequence2', dbxrefs=[]),\n",
       "  SeqRecord(seq=Seq('---CGATAT---'), id='sequence3', name='sequence3', description='sequence3', dbxrefs=[]),\n",
       "  SeqRecord(seq=Seq('AAAGGATAT---'), id='sequence4', name='sequence4', description='sequence4', dbxrefs=[])],\n",
       " '_per_col_annotations': {},\n",
       " 'annotations': {}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print the type of the align object\n",
    "print(type(align))\n",
    "\n",
    "#visualize a dictionary representation of the align object, with the attributes of this object as the keys of the dictionary. \n",
    "align.__dict__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We remove poorly aligned positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "alignment without poorly aligned positions:\n",
      "Alignment with 4 rows and 5 columns\n",
      "GGATA sequence1\n",
      "CGATA sequence2\n",
      "CGATA sequence3\n",
      "GGATA sequence4\n"
     ]
    }
   ],
   "source": [
    "#The best alignments are found at positions 3-8 in the sequence. We remove the poorly aligned positions outside of this range. \n",
    "sub_alignment=align[:,3:8]\n",
    "print(\"\\nalignment without poorly aligned positions:\")\n",
    "print(sub_alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'Bio.Align.MultipleSeqAlignment'>\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'_records': [SeqRecord(seq=Seq('GGATA'), id='sequence1', name='sequence1', description='sequence1', dbxrefs=[]),\n",
       "  SeqRecord(seq=Seq('CGATA'), id='sequence2', name='sequence2', description='sequence2', dbxrefs=[]),\n",
       "  SeqRecord(seq=Seq('CGATA'), id='sequence3', name='sequence3', description='sequence3', dbxrefs=[]),\n",
       "  SeqRecord(seq=Seq('GGATA'), id='sequence4', name='sequence4', description='sequence4', dbxrefs=[])],\n",
       " '_per_col_annotations': {},\n",
       " 'annotations': {}}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#the variable sub_alignment is also pointing to an object of type \"Bio.Align.MultipleSeqAlignment\"\n",
    "print(type(sub_alignment))\n",
    "sub_alignment.__dict__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Now, we will record the frequencies of each nucleotide at each position in the aligned set of subsequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One problem is that the raw counts (frequencies) of each nucleotide at each position would depend on the total number of subsequences sampled.\n",
    "\n",
    "To avoid this dependence, we can instead <b>record proportions or probabilities (ranging from 0 to 1)</b> of observing each nucleotide at each position. i.e. we need to divide (<b>normalize</b>) the raw frequencies (counts) of each nucleotide at each position by the total number of subsequences. This representation of the binding motif of a TF is known as its <b>position weight matrix (PWM)</b>.\n",
    "\n",
    "We notice that the first column contains base C twice, and base G twice. The second column always contains base G, and the third, fourth, and fifth columns always contain the bases A,T, and A, respectively. We can quantify this information with a position weight matrix (PWM). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Seq('GGATA'), Seq('CGATA'), Seq('CGATA'), Seq('GGATA')]\n"
     ]
    }
   ],
   "source": [
    "#We iterate through each sequence record in our sub-alignment and extract only the .seq attribute\n",
    "#This is the character string that encodes the DNA sequence from the alignment. \n",
    "aligned_sequences=[]\n",
    "for seq_record in sub_alignment: \n",
    "    aligned_sequences.append(seq_record.seq)\n",
    "print(aligned_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Seq('GGATA'), Seq('CGATA'), Seq('CGATA'), Seq('GGATA')]\n"
     ]
    }
   ],
   "source": [
    "#The iteration above can be condensed into one line. \n",
    "#We iterate through each sequence a in our sub-alignment and extract the seq attribute\n",
    "#This is the character string that encodes the DNA sequence from the alignment. \n",
    "aligned_sequences=[seq_record.seq for seq_record in sub_alignment]\n",
    "print(aligned_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#We use the alphabet defined above to make a motif object.\n",
    "from Bio import motifs\n",
    "m = motifs.create(aligned_sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Seq('CGATA')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Print the consensus sequence for the motif \n",
    "m.consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': (0.0, 0.0, 1.0, 0.0, 1.0),\n",
       " 'C': (0.5, 0.0, 0.0, 0.0, 0.0),\n",
       " 'G': (0.5, 1.0, 0.0, 0.0, 0.0),\n",
       " 'T': (0.0, 0.0, 0.0, 1.0, 0.0)}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print the position weight matrix for the motif \n",
    "m.pwm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.  0.  1.  0.  1. ]\n",
      " [0.5 0.  0.  0.  0. ]\n",
      " [0.5 1.  0.  0.  0. ]\n",
      " [0.  0.  0.  1.  0. ]]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIEAAACMCAYAAADr57kMAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nOzdd5xU1f3/8deZsjszW+hdKVawgxQRsUZFrGjsxh6NMfmanybfxG8SU/ya8k2MLcaKDXtBRAHFhqiAiihKVRCQDgss23ennN8fZ9ZdkLJldu7szvv5eNzH7LR7DrOX2bnvOedzjLUWERERERERERFp23xed0BERERERERERFqeQiARERERERERkSygEEhEREREREREJAsoBBIRERERERERyQIKgUREREREREREsoBCIBERERERERGRLBDwquHOnTvbvn37etW8iIiIiIiIiEib8+mnnxZZa7vs6D7PQqC+ffsye/Zsr5oXEREREREREWlzjDErdnafpoOJiIiIiIiIiGQBhUAiIiIiIiIiIllAIZCIiIiIiIiISBZQCCQiIiIiIiIikgUUAomIiIiIiIiIZAGFQCIiIiIiIiIiWUAhkIiIiIiIiIhIFlAIJCIiIiIiIiKSBRQCiYiIiIiIiIhkAYVAIiIiIiIiIiJZYLchkDHmEWPMBmPMvJ3cb4wxdxtjlhhjvjDGDEp9N0VEREREREREpDkaMhLoMWDULu4/Bdg3uV0D3Nf8bomIiIiIiIiISCrtNgSy1k4HNu/iIWcCT1hnFtDeGNMjVR0UEREREREREZHmC6RgH72AlfWur0retjYF+xYRL8QqILrVXcYr6y7jFdv9XHtZDtESiJZCrCy5VUC8PPmYSrBRsBawyctE3fX6twFg3GYM32XVxue22p/xgT8X/GHwRyCQl9zyIVjgtkABBCL1HpP8uf5ttdcDeRBsDz5/Wl9qaQJr3XFXvSm5FbnLmk1QswWiZe54jNUej+XbHY9VkKja8XG4zbGX3HzB5PFSe6xEwJ8HwXx3vAUKk8dcIeR2Sm6dISf5c077umNXWqVbp93K5KWTmXrJVApyC7zujohIi7PWcsUrV7CxfCOvXPgKAV8qThtFJBOk4n+z2cFtdocPNOYa3JQxevfunYKmRWS34lXJk+SiupPl6iKo2giVq6FyLVStd7dHt7oghwT4csD4cSfE9U9gLdhEcotBIg7Evfm3NYQJuM3nB3z1Tu7hu/Dpu39LjQuGgoWQ2wFyu0CoO0R6ucvak/v6m07wU8daiBZD+UqoSG5ly6D0Kyj/Fmo2Q02xC3YAfLnJ36up+x3Gq6kLE9PMl1Pv/w1g45CIui0QdmFRbgcI9YDC/SB/b4js6ba83u4YUwiZcf4545/c8t4tABz16FFMv3w67ULtPO5VK9S9O6xf73UvJBt16wbr1nndi1bFWsu1r13L43MfB+CsZ8/i5fNfJugPetwzEUkFY+0O85ptH2RMX+A1a+1BO7jvAWCatfaZ5PXFwLHW2l2OBBo8eLCdPXt2U/osIuDCnfJvoXy520qXQsliqFqXHBWxxY2EsPHkyXKA706WE1FIVLOTvFZ2xAS2Cx3qn+Ang6OcjhDqApHe0G4A5PWt20JdkwFUlkvE3fFasshtWz6H0q+hYpULJ7HudTY+F8rFq2izx6kvZ9t/a6LGjUYL94D8vtD+MHccFfaHgv3cyCNJq9um38ZfPvgLFdEK/MZPwBdg745788EVH9Ah3MHr7rUuev8TLzXgfEechE1wxYQreHHhi1REK/AZHyF/iBG9R/Dqha+SG8j1uosi0gDGmE+ttYN3eF8KQqBTgZ8Bo4FhwN3W2qG726dCIJHdiFVCxbdQttydNJctga0LoewbN4InVu6mpxi/CyLiFV73WOozgeTvx+cCt0TcBURjVnvds/SIV0HxPBf0bJ0Pmz91IWXlGjeNz/iT07JqvO5pBvO5aYrGuOlswUI3eqjjQGh/iAuH2h/sAkZJuVvevYXbZ95ORdS9t/qNn7iNk+PPoU+7Psy4agadI5097mUrohBIvKQQqEHiiTiXjL+EiV9N/O69z2d8JGyCcCDMsF7DmHLJFEKBkMc9FZHdaVYIZIx5BjgW6AysB/4ABAGstfcbYwzwb9wKYhXAFdba3aY7CoFEcFNXShZB8ZdQ/AVsXbCTkKfG1dWR1u+iNvhBNF4FW+a6oGfD+7DpIzeVyx8BEi7A8GqKVlvlD7lRRPEqNxKtw2HQ5RjoPBQ6Hq5gqBmstfzmrd/w70/+/d1JENSFQAA5vhx6FfZi5lUz6Zbfzauuti4KgcRLCoF2K5aIcd4L5/HG0je2ee+rDYEAwoEwA7sP5M1L3yQSjHjVVRFpgGaPBGoJCoEkq1jrToq3fO7CnqKZ7rJynQt6sC70aavTXrZhklPTauvz1BZ8rl97KFmnZ/tLG3dba9baQyCbcGHlxvdh/fRtAx8bVVjpJRNwgVC8ettgqOtR0Hm4u012yVrL/3vj//HQnIe2OQmCbUMggKAvSI+CHsy6ahY9CrQo6m4pBBIvKQTapWg8ytnPnc07y96hIrbte1/9EAggFAhxcNeDeeeyd8jP0TRlkUylEEgknWoDn82fQtEs2DAdts5L1uYJulERNuZ1LxvBQLDdzosiB9slV+LKr1udyx9xhXD94boRC/7cZNHcAN9fkYkdfEAzyfrN9Qo5G597fCKaHB1VXbfSU6wyufpTcrWyWJkrch0tgVjJtqtI1S+Une7pSK0tBKoNfdZPg9WvQtGM5O1xBT6twXfBUCVcoKl3u2Kt5bpJ1zHui3HfC4Dg+yEQQMAXoFteN2ZdPYs9CvdIV1dbJ4VA4iWFQDtVHavmjGfP4P0V71MZ+/7f9e1DIICQP0T/Lv157/L3KMwtTFdXRaQRdhUCaa0/kVQoWwZr34BvX4BNH9cFPtEytpkGk0knzYG8uqLF+X0h3Asie7gVikJdXZHjnHbucYnkylm1J0DGnwx2cprYuH/H6wo2+Ol+Fy41dJEKG3eBUSLKd8uA+wLgC7nRK9FSV0i7usitlFa5DipXQfmKZD2m5VC1NjkqKQuULIa1byr0aQtszAWhsksJm+DqiVfz3PzndhgA7UwsEWNd2TqGPDSEmVfNpG/7vi3XSRGRFKuKVXHKk6fw0eqPdhgA7fR58SoWblzIiEdG8P4V79M+1L4FeykiqaYQSKQpYhWw4T1Y9QqsnugCBHzbFmf2+oR5+5Anf1+30lB+P7dEtT/X9dEmXKATCO98X36/e3xrZfy7mAqTDJRCXYD96m5OxJKvTzLQ8+W6kKhipVvNauuCupXZWntIlIi7KYorX4Jvn08ez3h/DIukQTwR59IJlzJh0YRGBUDfPd/G2VC+gaEPDWXmVTPZu+PeLdBLEZHUqohWcOK4E5mzdg5VsapGP786Xs1Xm77iiIeP4MMrP6RTpFML9FJEWoJCIJGGsBZKFsKa191ony1zXHCw/UgfL/hyoN0B0O4g6DAQOg+DwgFuitbuQh5fQ4fSZCFfAHwF294W7u62TkN2EBIFofxbV/epaJabAlj8BVSu9ab/uxMrd6N9VjwDa6a42+IVrb/mkkgjxBIxLnzxQiYvmdykAKhWwibYVLmJYQ8P48MrP2T/zvunsJciIqlVVlPG8Y8fz5frv6Qq3vgAqFZNvIZlW5Yx7OFhzLhqBl3ztCiBSGugEEhkV4rnw6I7YNWE5KiIhKtBA94sbe0LuqWhOx4OXUZC5yMgr7erh2MM+PPB59v28dIydhQSFezjtl6nu+PFl+uOk+Iv3MixTR9709da8RpYOwW+vs/V+PHlunpJIlkoGo9y7gvn8uY3bzYrAKqVsAk2V25m+NjhfHDlBxzQ5YAU9FJEJLVKqks45rFjWLRxUbMCoFo1iRq+3fotQx8ayqyrZ9E9v3sKeikiLUmFoUW2ZxOwehLMv82dvNevheOFvD7QYxT0Phe6HOlO5H0BN91LWo9EzBWrzknzvHlrXZHyr+9zU70wECtNbx8ygfEl61x1cOGo8e94q78K3fZbtBRqNtUFwa1NaytK3oJq4jWc+cyZvLfivQbXwdhRYegdMRgKcwuZfsV0Dul2SHO72naoMLR4SYWhASiuKmbkoyP5etPXVMerG/ScHRWG3pGAL0D3/O7MumoWvQp7NberItJMKgwt0hDRElgyFhb+nztZj5V50w9/CLoeA73OgD3OdCet1kIwGfr4d1G7RzKXL5DeAKh6MywbB4vvcsWuE9Vta6pXbQHz3E6Q0ym5el0nCPeAcE/Irb2vg5sa6Q+71yAR3e5kYHcnBrUr05m6gug24YK0mq0uFKremCwmvrreCnSb3H1VRe52L0YOyg5Vx6o59elTmbFyRqMKoTaUxbK1eisjHx3Ju5e9y6Aeg1LehohIY22u3MyIR0bwzZZvqImn/m9SLBFjbelahjw0hFlXz6J3u94pb0NEUkMjgURsApY8CJ/9yv0cb/60gCbpNBQOuBl6nuJGGgTywef3pi/SepV9A/P/Csuf5HvFylsbXy4U7AuF/V1R8w6DoP2BENnTBTmJGrZd7S3XXXrFJtxIPVtTVyTcBFywW70RSr+CzXOgeB6ULoaSRS4sSheNBKIyWsnJT57MJ2s+aXQh1IaOBKovPyeft370FsP2GNao57VJGgkkXsrykUAbyzcyfOxwvt36LdFEtFHPbehIoFp+46dTpBMzr5rJXh32amxXRSRFdjUSSCGQZLctX8CMS6D8G1coN918QdjzHDjwd24FL19IwY80zaZP4Ms/wfq33WpftnEf8jyV29kVMy/sD+0Pho6DXPiT0zFZrNq61d3aQo0rm3CrC9pYcnRSFMqXuWBoyxwXDJUscmFeqleby/IQqLymnBOeOIG56+c2aSWcpoRAAHnBPN645A1G9B7R6Oe2KQqBxEtZHAKtK1vHEQ8fwZrSNY0OgKDxIVDtczqEOjDzqpns22nfRrcpIs2nEEhkR9ZMgfd/mCz47MH/g24nwPDHINjOTVcRaSxr3XH85R/ckvXxKjxfra4hwj3dlMeeo6D7D5JhT5ULebK51lW82r0Oxu9qGG36CFZNhA3TXH2y5oZCWRwClVaXcuzjxzJ/w/wG18HYXlNDIHBB0GsXvcaxfY9t0vPbBIVA4qUsDYFWl6xm+NjhrC1bSywRa9I+mhICgauP1j7Ung+v/JABXQY0qW0RaTqFQCLb+/ZFmHlpMgBKM+OHYQ+7Qs/ZfMIrzVP0MXx8DZQt9a5+VUPtKPRJ1ECw0OueZb5YpRvVZQLND4WyNATaWrWVox87msVFi5scAOX6c4nbeJNPogAiwQivXPAKP9jrB03eR6vWvTusX+91LyQbdesG69Z53Yu0+3brtxzx8BFsKN/Q5AA715+LxTa5hpDB0C63He9f+T4HdT2oSfsQkaZRCCRSX9VGeKWvR7VSDBwzEbodpwBImqb8W/j0F7D2dW9CzIYIdXfHuEKf1NsmFPoYVk+E9dNgy+fsdkRjFoZAWyq3cNSjR7Fk85Imn8T4jZ+Duh7EtYdfyy/f/GWzlpOPBCO8cO4LjN53dJP3Ia1UJoyEytLRMNlo2ZZlHDH2CIoqipo0igcgFAgxap9R9GnXh4fmPNSs977C3EKmXTaNgT0GNnkfItI4Wh1MpL4v/+TdKkm9z4WuxyoAksaLlsK8W+GrfydXuGr6iIQWkdMRev8Q9vmJK+K8fejjD3nXt7YkEAaSKwR2OxY6DXPHQrwKlj0Byx6H4i+97GHGKKoo4sixR7Ji64pmrYSTn5PPpIsm0bOgJ++teI+Jiyc2eVWximgFP3z+hzxzzjOc2f/MJvdJRGRnvt70NcPHDmdL1ZYmB0AGQ7e8bjw55klyA7l8uPJDPl/3eZNHQ5ZUl3DMY8fw9qVvM6TXkCbtQ0RSx+d1B0TSbvVEt1S0F/rfCMF8b9qW1mvtm2702lf/dqN/MiUA8oehzwVwwjswZjUMvB06DnSBj0b9pEcg7GqKhbrAfj+Hk2bCmcvhwN9CXh+ve+eZ9WXrGfrQUJYXL29WABQOhHnpvJfoVdgLYwyPnPkIPQt64jNN//hUGavkwpcu5IX5LzR5HyIiO7Jw40KGPTyMzZWbmxwAAYSDYV6/5HXycvII+AK8euGrFOQ0r35laU0pxz9+PDNWzmjWfkSk+TQSSLJPqlfcaYxmnDi0WeO7Q1U21YnwwUUNHIkWr4JPb4Rlj2XW1K/2h8D+v4A+57lRdbWBjxa285Y/B8hxIw0P/C0c9FvYMhcW3QErx3vdu7RZW7rWrYRTtqbZNXx+deSvOGGvE7a57fVLXuew+w+jPNr0FSUrY5VcNuEyquPVXHLIJU3ej4hIrXkb5jHykZEUVxc3az+RYISxZ4ylf+f+393WPb87Ey6YwKgnRzV5JCRAWbSMk8adxOSLJ3N0n6Ob1U8RaTqdkUr26TEKfDnetP3VPRDN8CK+6ZZVARA0ePWuLXPhtQGZFQB1PxFOmuFGm/S7xIUNGvGTmQJhN1Kr8xEw7CEYs8brHqXFqpJVDHloSLMDoBx/DkN7DeWWY2753n37dNyHcWPGEQlGmtNVKmOVXPPqNTz6+aPN2o+IyOfrPmfEIyOaHQCFA2EuOeQSLjjogu/dd3Sfo/n90b8nL9i8kgbl0XJOeeoU3v7m7WbtR0SaTiGQZJ9Db3UrdHlh+VOweTbEmv4NsmSBxf+GqcOhfIX3AZDxQd+L4PSvYeRL0Hk4BCJuOXdpHYKFbrpYG7e8eDlDHhzCurJ1zQqAADqEOvDSeS/tdNrXmAFjuHrQ1SkJgq6fdD0PzH6gWfsRkew1e81sjn70aEqqS5q1H7/xs2+nfbnnlHt2+pjfHPUbRvQeQa4/t1ltVUQrOOPZM3hjyRvN2o+INI1CIMk+4R4w+D73LXm62QS8exKsfUMjguT7rIU5v4LPf50MfzxeyaX7D+C0r2DI/VCwj6s9I5KBlm5eytCHhrKhoulLIdcKB8JMumgSHcMdd/m420+6nQGdBxBsZiBaGavkxqk3cvdHdzdrPyKSfWaunMlxjx1HaU1ps/eVl5PHpIsmkePf+Wh5YwzP//B5OoU7Nbu9imgFY54bw6uLX232vkSkcRQCSXba+zIY/oQ3QVAiCu+fAzMvccvVR5v/h1vagEQMZlwCX/8H4k1fhjUlCgfACe/C0ROgYG+FP5LRFhctZujDQ5u1FHKtSDDC7SffzuE9D9/tY2uLpeblNH+1x4poBTe/fTP/mPGPZu9LRLLD+yve58RxJ1KWgi8Vw4EwL577InsU7rHbx7YLtWPyxZObPRISXAh+/ovn89KCl5q9LxFpOIVAkr16/9CtapS/F/g9WLJ91SvwSh+Yc5Ob9hMt9bZotXgnVgHvjoJVE7wNgIwfDv4TjJoNXUa6mj8iGWzBxgUcMfYItlRuwTZz5FwoEOLUfU/lJ4f/pMHP6VHQgwnnTyAcaP4XChXRCv447Y/cOv3WZu9LRNq2t795m1FPjWpWgfpatUXwT9z7xAY/59Duh3LPKfekLAi69OVLeXbes83el4g0jEIgyW6dj4DTFsHBt4A/4pa2Tqd4JSx9yC3/Pf0sWD8N4tVQszV7AqFQN6974C2bgOljoOhDbwOg/L1h1Bzof1Oy5o+W+pLM9sX6Lzhy7JEUVxU3OwAyGHoW9OSxsx7DGNOo5x7T9xh+d/Tvml0sFVwQ9LcP/sbv3/k91no8HVREMtLrX7/O6c+cTkW0+Z8Zcvw5DOk5hD8c+4dGP/fKgVdy9oCzUxOCxyq48pUreXzu483el4jsnvHqQ8bgwYPt7NmzPWlbZIeqNsBX98HiO8HGIOZRzZ5ge1eLZc8x0PMUMEE3QiMFf2RlB55u3AlfSlxU733389+6Y87LAKj3eXDEI+DLBV/Au36INNCctXM47rHjKKlpXiHUWnnBPOZcO4f9Ou3XpOdbazn5yZOZvmI61fHqZvcnEoxw/ZDr+fsP/t7oUEoyVCb8HhUstnqvLn6V8188v1nLtNfXNa8rC69fuNsaaDtTFavi0PsPZcnmJc2ejgtuWtrdp9zN1YOubva+RLKdMeZTa+3gHd2nkUAitUJd4ZA/wNnrYeiD0O4ANzoo3f9NosWw8kWYcTG82BGmHgFf3AKbPk6OEiqGeFV6+yQtY+XL3gdA/f8fHPGom/qlAEhagY9WfcQxjx2TsgAoHAjz+FmPNzkAgmSx1HOfb/KJ1PYqohX855P/cMPrN2hEkIgAMH7h+JQGQA0tgr8roUCIKRdPScm0MHBTw254/Qb+/fG/U7I/EdkxhUAi2/PnQN8L4dT5rmbQnme7E+RAgRuVk25b58Oif8Ibw+ClTm7a2NzfwqpXoXINJGpcMBTzuJiwNM7Wha4QtJcB0MB/wCG3uulfIq3Ah99+yAlPnEBZTWpGakYCEa4ceCXnHHBOs/fVPtSeyRdPTsnUCIDyaDljPxvLdZOuUxAkkuWemfcMl4y/JGUBUCQY4R8n/oPBPXc4SKBR9uqwF0+f/XTK3vsqohX8+q1fc/vM21OyPxH5Pk0HE2mIRBy2zIHVk9wondKv3dSZWAas7BUshPaHQLuDoNMQ6DjYLeeNgUQ14HNTyZq5jHGb5dV0sLeOhQ3T8WwZ+AP+Bw68GYL53rQv0kjTlk/jtKdPS0khVHCrex3c9WA+uvojgv7UvT8+POdhbnj9hpTU6wB3snbuAefyyJmP4DP67q7V0nQwaaInPn+Cn0z6ScoCoFAgxOh9R/PiuS+mdLrpTVNv4v7Z96f0ve93I3/HzSNvTsn+RLLNrqaDKQQSaYqaLbDuLVg5Hta8DjbqlvhONL8WRMqEukFeX8jv6y4LB0DhfhDp7aa+1fbX+MCfxSGRFyHQD96Dd0/xbhRQz1PgqBc1AkhajTeXvslZz55FRQpHPLbLbceC6xfQs6BnyvZZ66KXLmLCogkp/db+zP3PZNyYcfhVtL11UggkTfDQnIe4YcoNKXsvMRj6tO/DvOvmkZeT2hVAY4kYIx4ZwWdrPyOaiKZkn5FghF8O/yV/Ou5PKdmfSDZRCCTSkqyFrfNgzRT49gUo/gJ8Oe4+r4pL75ZxQdB3IVE/93O4h7s9tzPktIdAoQuJEtUuNMK65/pywJ/r7mstEjUQr3FFv8H13ZcDz3lQcLv9wVD8ZfrbBRcCnjoPggXetC/SSJO+msR5L5yX0gCothbGcf2OS9k+66uMVnLo/YeydMvSlBRLBXcyNGqfUTz3w+cIqH5X66MQSBrp3o/v5Vdv/iplARC4IvifXvMp+3feP2X7rG992XoG3DuALVVbUrbPSDDCz4f+nL+e8FcVyhdpBIVAIumUiEPJItj8KRTNgI0fJqePBQGTDIZa0QcxfxhyO0FOJxcOhTq7y9wubgsWQCA/WTcpz40u8YeTo4tCLizy5bjN+FwIYxOATb4Mlm1fj2TQBNteGgCfWynN+OpGMiWqXcHseBXEK93omlgFxMrdax0rhZqtULUeqouS26a6n2s2p+uVrBPI9y4gPOo52GNM9o78klZlwqIJXPTSRSk9CYoEI/xmxG/4/TG/T9k+d2Tp5qUcdv9hlEVT9389EoxwQr8TeOm8l1I6hU3SIBNOXhUCtRr/mvkvfv/O71MafkeCER4981HOO/C8lO1zRz749gNOGndSyt+3rxl0Df86+V8KgkQaSCGQiNdqg6HiubDlMyia5QoDR0tcvZ5EzNsCweny3dSz3OQHYl+9Sx8u7PElQ6JEvbAoUbfFKzNr2l2TGDwJAtsdBCd/pGlg0io8P/95Lp9weUpPJHL8OYzsPZKpP5qalvo6ExdP5MKXLkxZjQxwJ0Mje49k4oUTyfHnpGy/0sIy4cRVIVCr8NcP/sr/Tv/flL9vXH7o5dx76r0p2+eu/N+H/8ef3/tzymq4gRvFdOmhl3Lv6HsVBIk0gEIgkUxVs9VNJSv+EjbPcUFR+QqoWgcYN4oG3CiXRI2nXZU24Minofd5oJoikuGenPsk17x2TUoDIIDued1ZcP0COoQ7pHS/u3LjGzfywKcPpPSELhwIc8QeRzD54smEAqGU7VdaUCactCoEynh/nPZH/jHjHyl9v/AbPwd3PZiPf/xx2kYQWms59elTeWfZO1THU/fFXV4wjwsOuoAHT39QhfJFdqPZIZAxZhRwF+AHHrbW/m27+48FXgGWJW8ab6398672qRBIZBesddOUypdD2XJ3WbIIShYnQ6L17nEKiaQxztkEuR297oXILj362aNcP/n6lAdA4UCYD6/8kIE9BqZ0v7sTS8Q4cuyRfLbuM2KJWMr2Gw6EObzH4Uz90VTCQQ9qm0njKASSXbDW8j9v/w93f3x3SgMgcEXw5/90Pr0Ke6V0v7tTUl3CAfcewJrSNdgUjn6OBCOcPeBsHj/rcQVBIruwqxBot5UFjTF+4F7gRGAV8IkxZqK1dsF2D33fWntas3srIu7DYm4nt3U8/Pv3W+tWKCtfXhcUlS6Gkq+geqO7L7rV1capX48H66amJarrCiRLw/lyd/xagquNFGzvfmeRPdxqbAV7u4LbeX0hsiesmgAfXe3qFKVb4f6gqSOS4e7/5H5unHpjygOgSDDCv07+V9oDIHBL0U+8cCIH3HtASoulVsYqmb12Nsc9fhxvX/p2ylf6EZH0sNZy09SbUj5iEFxYPP788WkPgAAKcwuZcvEUjhh7REr/XRXRCsYvHE91rJpnznlGKyaKNEFDlpcYCiyx1n4DYIx5FjgT2D4EEpF0McaN6MjtCB0H7fxxibgLhL4riFwENcmiyJVroWKNm3pWWyA5Wupq7vhzwQTq6vTUshZXqyfutkSUVlHk2gSSBaWTG2bbb2Vr6w0lqgELgQK3OlpOR7daWriHC3ZyuyTDuc7bbg2tsVOzxbvwrXBAcoU3kcx010d3cfNbN6c8AAoFQpy+3+lcM+ialO63Mbrnd+eVC17h5CdPTum/rypWxdz1czn6saOZdtk0CnK16p9Ia2Kt5WeTf8Zjcx9LeQAUCUb47yP/m+P7HZ/S/TbGwd0O5t7R93L95OtTHgRN+noS5zx/Di+e96JWTBRppIb8j+kFrKx3fRUwbAePG26MmQusAX5prYBe1pQAACAASURBVJ2//QOMMdcA1wD07t278b0Vkcbx+d1qXqHOu32otVBaCkUbYxSv30zl1mJqKiuJVVcSq64gXl1JvKYCG60gXlNJIlpBIlqJL1FGkBJyTCk5vjJy/eWE/OXkBioIBSvJDVQSClbi98UxxmKwGGPxmcQ217/bsFhrsBis9WEtJKwved1tCevDWkNNLIeqWJjqaJiqeJjqWB7V8Tyq4/nUJAqIUkDMFIA/gglG8OdE8OWE8QfdZSA3QiA3TDAcIScSobBzZzp0yaNTZ0NOSwyaiVcli117wB8CNGz6e8Z3r5te2eb54KK4153YoX98+A/++N4fUx4AGQy9CnrxyJmPeF5IdGSfkdxyzC3cOv3WlJ4MVcWqmL9hPiMeGcH0K6bTPtQ+ZfsWkZaTsAmuefUanpn3TMoDoBx/DsN6DWvxVRAb4vLDLufdZe/ywoIXUvoeXxGt4M1v3uSMZ85gwgUTVChfpBEaEgLt6FPT9l/9zwH6WGvLjDGjgQnAvt97krUPAg+CqwnUyL6KSANZC+XlUFT0/W3jRli9GtauhQ0bYPNmKC52j/f5IDc3gN/fFeiKtW5fiQTE4xCLua018/kgEAC/320+X92gIGshGoWqKsjNhcJC6NABOneGbt2gRw/o2dNd337r2NHtd5cC4eToKg/EKgCPAqhMljUBEGTq7//W927lbx/+LeUnQeC+CX/9kteJBDNjRbxfj/g105ZPY9ryaSktllodr+arTV9x5Ngj+eDKD+gYVu0vkUyWsAkue/kyxi8a3yLvfR1CHXjpvJcypmbOA6c/wEerP2LJ5iXEbeq+jKiIVjBt+TRGPzWaSRdNIjeQm7J9i7RlDQmBVgF71ru+B260z3estSX1fp5sjPmPMaaztbYoNd0UkVrWwpYtsHx53bZ4MSxaBCtXuvvKytxjc3Prgol4HGpq3LYz8bgLQdqyRGLXr0Gtqiq3bdjgXt9afn/d62qM2180CtXVEAq54KhrV9h7bzjgANhrL+jb1229/Z0JmiCQ2tEODVI819UzEskQ1lp+/+7vuWPWHS1yEhQOhHlizBPs03GflO+7qYwxPPfD51yx1LI1u39CI1THq1m6ZSlHPHwEM66aQefI7keAikj6xRNxLhp/Ea999VqLvfdNumhSWldB3J1QIMSUi6dw6P2HUlqT2rqIlbFKZqycwUnjTuL1S15XoXyRBtjt6mDGmADwFXACsBr4BLio/nQvY0x3YL211hpjhgIv4kYG7XTnWh1MZMd2FfKsWAHr1rnH5Sa/7KiqalioIelnDEQiLjCKxyEvsJF19+3h3UpuY9ZCuLs3bWeqpzNgxZ50uigzBuEmEgl+/OqPeXre01TFqlK+/1AgxJj+Y7j5qJtTvu9UWFy0mB+9/COq4qn/twdMgG753Zh++XT26rhXyvcvTaTVwQSojlUz+qnRzFg1o8Xe+24afhPnH3h+yvedCtNXTOeXU3/ZIu99uf5c9u+8P9Mum5ZRAZiIV1KxRPxo4E7cEvGPWGtvM8b8BMBae78x5mfAdUAM9xX3jdbaGbvap0IgyXY1NS7c+fJL+OwzmD0bli1TyNPW2Ql7Qfkybxof+hDsdRn4gt60n4kUAqVdIpFg4AMD+WLDFy3Whs/4yM/Jb7H9p0JNvKZFTgJr+Y2fj3/8MYN67GLxAEkfhUBZr6Kmgr539WVjxcYWayPgC2TM9NedqYpVURNvuQ+2kUCEr3/+NT0Le7ZYGyKtQbNDoJagEEiyyfr18OmnLuyZORO++MLV5AknR6yWlbXdz2bBoBsNEw7XXebk1NXi8fm2/Rnq6hBtfxmPu1CsogIqK+suW9NrZz/5L/j6Xm8KROf1gVMXutpE4igESqtEIsGB9x3IoqJFnvYjW/iMjxlXzmDYHjtaz0PSSiFQViupKqHfXf3YXLXZ665khVAgxOKfLaZ3Oy1EJNlrVyGQ1tMTSbHawOeTT2DaNJg714UVoZArvly/sHJpaqdFp5TP54od1xY+7tSp7ufaAskdO0JeXl24Ew67EUy5uS7oCSYHnNQWlE4k6kKdpjCmLizy+900q0DA7bumxtXlqalxwVBtWFRR4V7ndetcQeyNG2HTpu8XzK5I/bT8HetzASwdC/F0NVhP+QpY9jjsdXlytTCR9IklYvT/d3+WblnqdVeyRsImGD52ONOvmM5RvY/yujsiWWlL5Rb63dWPrdVbve5K1qiKVbHvPfuy6PpF9OvQz+vuiGQcjQQSaaayMhf2TJgAr73mVtrKzXWhQqaupNW1a12x4r59YcAA6NMHunRxwU67di7Qqa52RY8TCRe+1BZFDrbS2US1RaFratzP4EKk2ql3paXu97dpkwuNFi2CpUvrajOtWOHCpeawFnjzaNj4Ad9faDENcjvDaYsgt1P6285EGgmUFjWxGvb99758u/VbT9rPdgbDmz96kxP2OsHrrmQvjQTKSkUVRfS7qx9lNWVedyUrBX1BvrzuS/bvvL/XXRFJO00HE0kha2HePJgyBZ5/3tX0CYVcgJApn6/atYP999825NlvP+jd2wVA8bgLeHw+1/ccLRq1Q4lEXU0mn8+NeCovhzVr4JtvYMGCupDom2/cz4ndzPKyFtjyOUw9EuIerBIG0GkInDANApldNyAtFAK1uKpYFXvftXfKV8OSxjEYXrvoNUbvO9rrrmQnhUBZZ13ZOva+e+8WWQFMGi7gC/D5tZ9zYNcDve6KSFopBBJppi1b4K234OWXXfgTjbpRPtXV3vbL73cBz6GHwsCBMGyYu15Q4KZEGaOQpyXVD4kCAfc6r1gBn38Os2a5gHDOHDeyqNZ3b7nTx8Dq18B6NFys32Uw5F4I5HnTfqZQCNSiKmoq6Hd3PzaUb0hru7JjBsP488dzVv+zvO5K9lEIlFVWlaxi33v2bdHi79JwfuNn9o9nc1iPw7zuikjaKAQSaYJo1IU+f/+7G/mTm+t9DZ+994ajjoIjj3SXe+9dF0Tl59cVVhbvVVW5LRx2x81nn7lpg7fdlnxAxSqYdDBEi73r5L7XwcB/ZveIIIVALaaspoy+d/ZlU+Wm3T9Y0uq5Hz7HeQee53U3sotCoKyxfMty+t/bn+q4x98UyjZ8xsesq2YxpNcQr7sikhYKgUQaYdMmuO8+uOMOFwR5GfxEInDccXDmmXD66VBY6KZyFRR41ydpupqa7UZlbfgA3j3Ju2lhAHueDcOfyN4RQQqBWkRxVTH97uxHcbWHIafs0rizxnHJoZd43Y3soRAoK3y96WsO/M+BRBNRr7siO+AzPqZfPp0RvUd43RWRFqfVwUQaYP58+Otf4aWX3Ge1So/Oyw88EE45Bc49103zqqpyo3z8fm/6I6nzvWl5XY+CgbfDZ7/0ZrUwgJXjoWojHD3BjQjKtlXDQt2gar3XvWhTiiqK2OuuvSityeDlD4UfTfgRVfEqrh50tdddEWkT5m+Yz2EPHEYskaGrgggJm2DkoyN569K3OL7f8V53R8QzGgkkWW/rVvjVr+DJJ91IjXg8/X0IheCii+B//ge6dXOBTzic/n6IR2ZdCSue8y4IAsjpCMPHQdejIZjvXT+yTbpHIrXwSKANZRvY6+69KI+Wt2g7kjr3jr6Xnw75qdfdaPs0EqhN+3zt5wx+aDBx68GHSGk0g2HyxZMZtc8or7si0mJ2NRJIFUQkq738sls9a9w4N/In3QFQ9+6u5tD69XDnna7GT36+AqCsM/Qh6HM++D2szVOzGd47FWb/DKJlEFcxS2mcNSVr6HtXXwVArcz1k6/njpl3eN0NkVZr9prZCoBaGYtl9FOjmbh4otddEfGEpoNJ1rr9dvj9772Z9uX3w3/9F9x6q/s5lGUzcGQ7Pj8MGwvhPWDR7d6OCFr2OKyZDIf9DfpcAL5c1z+RXSiuKmafe/ahMpaeN1SDwWTCyIoWlrCJtLRz49Qbyc/J58eH/zgt7Ym0FUs3L2XYw8PS9n/Vhw/a/ltfWl5Pi+XMZ89k+uXTGdlnZIu3J5JJFAJJVrr1Vvjb37wJgAoL3XLzAwZAXpbW4pUdMAYO/TOEe8BnN3lbLLp6I3x0FSz8Jwz5D3Qc7EYpafk52YHNlZsZ8ciItBRCDfqCDOk5hNP2O63F28oE68rWcd/s+9Ly2t4w5QZ6FfZi9L6jW7wtkbbgmy3fMHzscNJRWiPoC3LqfqcytOfQFm8rEyzetJinvnwqLfWVTn36VKZdPo1BPQa1eFsimUIhkGSdJUvgL39xBZfTLScHZs6EvfbS6B/Zif2ug8geMONiFwRZDwtMliyEt4+DzkfCIX9yl76g20SAjeUbOfKRI1lRvKLFP6wHfAEO7X4o713xHgFf9nx8GdRzED+d9FMqoi07QrAyXsm5z5/LU+c8xVn9z2rRtkRau682fcWRY49kS9UWLC0bAoUCIc7a/yyePufprBgBWat7fnfu+fieFn/vK60p5djHjuWtS99iaK/sCNlE9LWuZJ3f/AZiHp1XX3cd9O6tAEh2Y4/T4bSF0GkI+DNguFjRDHjnRHh9EKx41oVTMQ+nrElGWF+2nqEPD2VF8Yq0jFQpyCng1QtfzaoACOCyQy/j3APOJRxo+WJxFbEKLnrpIp6f/3yLtyXSWi3cuJAjHj6CzZWbW3zaksGwR8EePHLmI1kVAAHcdvxtHNb9MHJ82y+tmnqlNaUc//jxfPjthy3elkgmUAgkWWfmTO9CoMsvd4WfRXYr0gtO/AAO+TP4w2REEYCSxTDzUnilL8y7FSpWQ7QUtBxu1llTuoYhDw1hVcmqtARA4UCYVy54he753Vu8rUx0/2n306d9H/ym5etzVcYquXzC5Yz7YlyLtyXS2ny5/kuGjx2elhFAAJFghCmXTCEczL4VQ/w+PxPOn0BBbkFa2iuPlnPykyfz3vL30tKeiJcUAknWyWn5LxR2autW79qWVsj4YMCNcPLHULAPBDJgVBBA1QZY8DeYsAe8Owq+eRSiJW5LpKc4pnhn5daVDHlwCGtK16SlXkMkGOGWY27J6sKdoUCIKRdPIRJMzwqClbFKrn31Wh757JG0tCfSGny29jOOevQotlan58NcJBhh3Jhx7NNxn7S0l4m65HXh1QtfTctISHBB0OinR/PWN2+lpT0RrygEkqzzox95twT7X/4CZWXetC2tWPuD4NQFMPB2CLZLjgzKEEUz4ONr4MWO8MG5sOoliFVCzVaIV3vdO0mxZVuWMeShIawrX5eW5ZBz/bmM7D2SX4/4dYu3len6tu/LM+c8QySQviDoZ5N/xn2z70tLeyKZ7OPVH3PMY8dQUl2SlvYiwQhXD7yaMQPGpKW9TDZ8z+Hcdvxt5AXT80VYRbSCM589kylfT0lLeyJeUAgkWedXv4KgR3Vtp06FceMUBEkT+AKw77Vw1krY72cuCDIZVBvFxmHtVPjgPHixHUw7Beb9GYpmuTBIoVCrt2TzEoY+PJSNFRvTthxyp3Annvvhc1lXC2NnTt3vVK4fen1aRwTd9MZN3DXrrrS0J5KJZqycwfGPH09pTWla2gv4AgzoPIDbT749Le21Br844hcc3+94Qv70FNWsiFZwzvPnMHHxxLS0J5JuCoEk6xQUwGuvebc8+09/Cv/3f1BRAfGW/yJd2ppgAQz8PzhtEfQ6PRkGtXydkEZJRKFoJsz/C0wdDi8UKBRq5RYVLWLoQ0PZVLEpbQFQJBBh8sWTaRdql5b2Wou/nvBXDul2CME0rdJXGavkf975H/7+wd/T0p5IJpm+YjonjjuR8mh52trMz8nPyiL4u2KM4amzn6JLXhdMmmokVsYqueDFC3hxwYtpaU8knRQCSVYaORKmT4d27bwZFXTrrTB4MMyZ40YFKQySRsvrDUePd/WC9jwHfCFIwwoaTdKQUChWDjXFbiqZZJR5G+Yx/OH0FUIFNxXintH3cGj3Q9PSXmvi9/l55YJX0lYsFdy34n+e/mdufe/WtLUp4rW3lr7FKU+d0uJLlNcXDoSZcP4EehT0SFubrUVBbgFTLk5vkezKWCWXvnwpT3/5dNraFEkHhUCStQYNggUL4OSTIZKekfXbWLgQhg6Fk06CyZOhshKqqtLfD2nl2h8ERz0HZyyBfa8Df8RtmWz7UOi5MLx2gJtK9vmvYekjsPlTqNkCiRo3aihaCok2kJbGKpJhV0WrGAk1d91cRjwyguLq4rS1GQ6EObv/2Vw58Mq0tdnadM3rmtZiqeCCoL99+Dd+987vsDY9YaCIV6Z8PYUznj0jrQFQJBjht0f/lmP6HpO2NlubA7seyIOnPZi2KbHggqCrJ17No58/mrY2RVqa8eoP+eDBg+3s2bM9aVtke6+/Dlde6UbllKZnyvf39OoFN9wAV1wBoRD4fN6EU9LKRUvg6wfc6l2JKMQ8OqBTJVAAhftDYX9odwB0HASFAyDcE2zChUQ2uUKV8YEv6EZFpbuGTCLqQp3v9SXXhVllS2Hz51A8F0oWQcliqFyd3j4CXNTwv/mz18xOax0MAJ/xsU/HfZj7k7mEAump/dCa/XPGP/nDtD+k/UT1usHX8Y8T/6FaTY2RCa+VwrsGmbhoIhe8dAGVaRyZmuPP4Zg+x/DGJW/o/1UDXDXxKp758pm0/o7CgTB3jrqTaw6/Jm1tijSHMeZTa+3gHd6nEEjEicfh1Vfhf//XjRCqqfFumlb//jBqFJx3nhuxVFUF+fngz7DSL5LBElFYMwUW3wUbP3QnIPE2NtQsWAg5nSC3/tbZbeGeEOoOoS7uMTntIFAIPr8Lj6wFEoBN/ly7mbrN1F766q7Hyt3IpJrNUF0EVeuhcg1UbXDXqzdBzSZ3Wb3JPS4Nq2g1WANDoFmrZnHiEydSFk1vFfv8nHzm/mQue3XYK63ttlbWWk5/5nTeXvY2VbH0/f/OC+ZxxWFXcPcpd+uEtaEy4XVSCLRbLyx4gcteviyt4QJAj/weLLh+Ae1D7dPabmtVHavm8AcPZ1HRorSsVFkrHAjz9xP/zs+H/jxtbYo0lUIgkUb64gv4+99h/Hj3ua3SwzIl4TAcfTSccQaceSZ07uz6E4lAToaWgJHvKyuDRAIKCz1ovHItfPM4LLkfqoogUW/ESrbxBcEEXTFtX8Bd1t9swoU222wxd9kWQrQGhEAffPsBo54cldZCqOA+XD97zrOc0f+MtLbb2pVUl3DgvQeyunR12mo2gRsRdPHBF/PAaQ8oCGqITHiNFALt0tNfPs3VE69OewAUDoSZcdUMDut+WFrbbe1WFK/gkPsPoaS6JK3tRoIR/nTsn/jlkb9Ma7sijaUQSKSJiopg7Fh4+mlYtMhN0ypJ79+a7+ncGQ4/3BWWPuYYGDjQrXhWWelCIU0h81YsBuXlbjpfbi4sXQozZsCHH8Ls2S5g9Iy1bjrS0rGw/BkXBsWrwUY97JSk1W5CoHe+eYfTnz09rdOLoG6K0T9P+mda220r5m2Yx7CHh3nye/vhgB/y6FmP4jMqM7lLCoEy2mOfPcZPJ/807QFQJBjh7lF3c9Wgq9Lablvx+pLXOfu5sz35vd181M387ujfpbVdkcZQCCSSAlu3wjvvwMsv1xVyTiQyo5hzhw5w0EFw8MEwZIgLiPr2daFQVZXrZ26uG1UkzReLud9/PO5Wl8vNhY0bYf58mDUL5s6FL7+EJUu+P6UwYz6D1wZC346HFc9AxSo37Sme3pNISbNdhEBvLHmDs58/O+1BQsAXYFD3QXx41YdaErkZnvziSa597VpPgqAz9juDJ89+Er9Pc5Z3SiFQxnrw0wf5xeu/8GQE0JgBY3jq7KfS2m5b89u3f8udH93pyXvfTcNv4s/H/Tmt7Yo0lEIgkRSz1o0MmjIFXnjBLfUeCrmi0pn0GSs/H/r0cYFQ376w774wYAD06wc9erjwojbMqg0zAjoHo7rabfVfl6IiWLkSvv7a1YxavrxuW7vWPbYhMun42Eb5Slj9KiwbB1vmgD8E0TJc3RxpM3YSAr22+DXOe/G8tJ8EAXQIdWDh9Qvplt8t7W23NVe+ciXPznvWk2/FT977ZJ4/93kFeTujECgj3fPRPfz6rV+n/f+Mz/jYu8PezP3J3LQued4WxRNxjn3sWD5e8zE18Zq0th0JRrh+yPX8/Qd/17RYyTgKgURaWEUFvPcevPKKGy20fLkLheJxd1+mqg2JakOhTp2gZ0+3devmpp516ODq2OTkuGAkGnWfI41xgZHP5y6DQa//NdtKJFxf4/G6DVxx7Zwcd1leDsXFsHkzbNjgwpzVq92ono0bYcWKxoc8u9MqPoNHS2DDdFg71RWXrlgB/rBCobZgByHQywtf5uLxF3sSAIUDYab+aCpH9T4q7W23RdWxagY9OIjFRYvTWiwVIBKIcHy/4xl//niC/gz7g5AJMuEEsVX8AUqff874J7e8e4sn730qgp9aRRVFDLh3AEUVRWlvOxKM8ONBP+aOk+9QECQZRSGQSJrV1MC8ea4GzAcfuClCtcEQuCLBre2zWDAIHTu6YKh269DBTTGLRNxlYaGrT1RQAHl5dVs47LZQyG21gZHPV3dpTN2WSC7etP2WSLggqqrKbZWVLmQrL3dbaal7bUtK3GX9+4uK3LZpk7v0qrZTa/u9A1BTDBs/SIZCrydDoRDEKrK3wHRrFOoGZ6/b5qbn5j/HFROu8OQkKBKM8Idj/sB/j/jvtLfdlq0oXsHB9x1MaU1p2tuOBCMc1fsoJl4wkdxAbtrbz2iZcHLYKv8AtYzb3r+Nv7z/l7RPIQL3/+Tps5/mzP5npr3ttuyjVR9x3OPHefb37NJDLuU/p/5HQZBkDIVAIhmgpsZNIfvyS/jsMxcMLVzogotw2NWZyeRRQ5IabeIzeE0xbJwBmz6GDe/Bls8hXpkMhsoVDHklkO8uEzVQsB90GQGdj4TOw6Fgn++dhI6bO45rX7vWkw/Muf5cju93PJMumqQPzC1gytdTOOf5czwb3TWs1zCmXDKFUCCU9vYzViYc523iD1Dz3fLuLdw+83bPAqBrDr+GO06+I+1tZ4O7PrqL377927Svbgnud3vegecx9oyxKpQvGaHZIZAxZhRwF+AHHrbW/m27+03y/tFABXC5tXbOrvapEEjEKS52o4bmz3eFhBcuhG++cdOSystdQOT3uxDJy6XqJTXa7GfwynWw+VPY9Em9YKjCTSNLRFVwOpV8Oe51tTH32tYPfDoeDoX9YTcFeh+e8zD/NeW/PAkJAHoV9GLB9QsozC30pP1s8Ju3fsM9H9/jyYluOBBmUI9BTP3RVCJBLVkJKATKANZafvPWb/j3J//25P9FwAQY2GMgM66aodpZLcRay5jnxvDG0jeoiqV/5ZZIMMJZ+5/FE2OeUKF88VyzQiBjjB/4CjgRWAV8AlxorV1Q7zGjgZ/jQqBhwF3W2mG72q9CIJHdq6yEb7+tK0BcGxItW+ZCorKyupAoGtVIokwRCLjfi8/npq/F49Cli6u/NGOG171Lo6qNULLIbcXzXMHp0q+hugj8EcC4cEgjh3bAQCACJuBGWflyIK8ftD8YOg6Cwv1d2JPXb7eBz/b+8/F/+OWbv/QsAIoEI8y6ahYHdzvYk/azRTwR5+jHjuaT1Z8QTUTT3n4oEOLgrgfzzmXvkJ+Tn/b2M45CIE9Za/nFG7/g4TkPexIAgSuCv+D6BXTP7+5J+9mirKaMA/9zICu3rsSS/mM+L5jHqH1G8ewPn1XYJ55qbgg0HPijtfbk5PWbAay1f633mAeAadbaZ5LXFwPHWmvX7my/CoFEmq+qatuQaPlyWLUK1q1zhY43b3ZL25eVuWCitiAyuGCittCz7JwxdXWMjHGfoaNR99oFg64OUocOrqh2t26uqHbv3nUrsvXtC127Zsbn/4wRr4bSJcmAaKG7LPvGLVNftdGFQv6wW7K+rY4i8uW46XPgpm8lopDTAcI9Ia8vFO4HhQNc0FO4P+R2Skmzd8y6g9+9/TsqYt68ppFghHtPuZfLB17uSfvZZmP5RgbcO4BNlZs8aT/kD9G/S3/eu/w9jfrKhD8CWRoCWWv56aSf8sQXT3gWAIUDYd645A1G9hnpSfvZZuHGhQx+aLBnv+9IMMIP9voBL577ogrli2d2FQI1JJ7sBaysd30VbrTP7h7TC9hpCCQizRcKwX77uW1XrHVhUG1R5Prb+vVuVNHate56bXBUu3R8IFC3Clj9As61+60t2ByLuWAp5vGgjtrVyvx+t9X2uVZtfxOJuv7WLgVfUADt27ui11271q2U1qWLu61Tp7qi2J06uaXjG+vOO93UPi8cckhdcXJv5QIHJreknOQG+BNbCcVXkpvcQvEVhGNLCCbWk5PYSCBRjN+W4LcVWAIkTA4Wl24a4hgbd5fEMC24mpnFYAm4to0fix+LwZDA2Bg+qrEEifkKiJn2xHydqPF1ocrfh6rA3lT5e1Pt35Nq/57U+Lq50MvtGLYmtxR6Zfl13L3iARIefDMKEPLlcHrPozg9ty+bFkzzpA/Zxgc8OfxPjHnvJqri1WlvvypexZfrv+SYO4dxmz2VQDOmR8SiNVSWNu2E7shzm1eA1/h8BHOb9+aZmhi3ebLx/521lksn3czUyo+Je/TeF/bn8ou9z2OfolLWFk1u9PMDuSF8PtWYaYyuwJ2DbuSG2f+g0oP3vopoBZO/msSoh0bw1FF/IdiMEUGbVq9hzderUti7hsvf81AGn36yJ21Ly2rIEbmjry62fxdtyGMwxlwDXAPQu3fvBjQt0jL+eKzhj+953Yv0MUD75LZ3I54Xx0dlNExFNEIlYSqIUEHdzzu7rYQCSpNbGQWUkUeUHBIYLD4SGBL46133bXOfu6Xukb5tHuWe6SNBmEryKSM/2VIhJeQlyonUuF6Fqdzmcke3hakkl2pMDbApuS1tgV9CUjF/4I/8qeUayCYGTLgGX0ENFAD5yS2n3pYLhIBw8jK36gtFBgAABVtJREFU3mZ2sIH761V/iwLVya0quVWAqbaYmijURKHG3UZZcit1lyZeTQ7V5FAELGnRl2N3nro2Qn+/dyMB9tpSw2N/nkooNtWzPmSjUcC/B8K/hnvVgzidq8sYffztXnUADn/Au7ZrdevmvnXxsP1OBxxbd/2LP8K87PhbFFoH+6c/B/jO8JXV3Pb7xzE87l0nstCPgRXHwcsDvOpBnOiyT+jy4VX4x61o8l46HQD7nZjCbomg6WAiIiIiIiIiIm3GrqaDNWRs4SfAvsaYfsaYHOACYOJ2j5kIXGqcI4CtuwqAREREREREREQkvXY7HcxaGzPG/Ax4A7dE/CPW2vnGmJ8k778fmIxbGWwJbkD8FS3XZRERERERERERaawGVamy1k7GBT31b7u/3s8WuD61XRMRERERERERkVRRqXkRERERERERkSygEEhEREREREREJAsoBBIRERERERERyQIKgUREREREREREsoBCIBERERERERGRLKAQSEREREREREQkCygEEhERERERERHJAgqBRERERERERESygEIgEREREREREZEsoBBIRERERERERCQLKAQSEREREREREckCxlrrTcPGbARWeNK4CHQGirzuhGQlHXviFR174iUdf+IVHXviJR1/4pU+1touO7rDsxBIxEvGmNnW2sFe90Oyj4498YqOPfGSjj/xio498ZKOP8lEmg4mIiIiIiIiIpIFFAKJiIiIiIiIiGQBhUCSrR70ugOStXTsiVd07ImXdPyJV3TsiZd0/EnGUU0gEREREREREZEsoJFAIiIiIiIiIiJZQCGQZBVjzChjzGJjzBJjzG+87o9kD2PMI8aYDcaYeV73RbKLMWZPY8y7xpiFxpj5xpgbvO6TZAdjTMgY87ExZm7y2PuT132S7GOM8RtjPjPGvOZ1XyR7GGOWG2O+NMZ8boyZ7XV/ROrTdDDJGsYYP/AVcCKwCvgEuNBau8DTjklWMMYcDZQBT1hrD/K6P5I9jDE9gB7W2jnGmALgU+AsvfdJSzPGGCDPWltmjAkCHwA3WGtnedw1ySLGmBuBwUChtfY0r/sj2cEYsxwYbK0t8rovItvTSCDJJkOBJdbab6y1NcCzwJke90myhLV2OrDZ635I9rHWrrXWzkn+XAosBHp52yvJBtYpS14NJjd9+yhpY4zZAzgVeNjrvoiIZAqFQJJNegEr611fhU6ERCSLGGP6AgOBj7ztiWSL5FScz4ENwJvWWh17kk53Av8NJLzuiGQdC0w1xnxqjLnG686I1KcQSLKJ2cFt+kZSRLKCMSYfeAn4hbW2xOv+SHaw1sattYcBewBDjTGaDitpYYw5Ddhgrf3U675IVhphrR0EnAJcnywLIJIRFAJJNlkF7Fnv+h7AGo/6IiKSNsl6LC8BT1lrx3vdH8k+1tpiYBowyuOuSPYYAZyRrM3yLHC8MeZJb7sk2cJauyZ5uQF4GVeWQiQjKASSbPIJsK8xpp8xJge4AJjocZ9ERFpUsjjvWGChtfZfXvdHsocxposxpn3y5zDwA2CRt72SbGGtvdlau4e1ti/uM9871tpLPO6WZAFjTF5yIQaMMXnASYBWh5WMoRBIsoa1Ngb8DHgDVxj1eWvtfG97JdnCGPMMMBPY3xizyhhzldd9kqwxAvgR7lvwz5PbaK87JVmhB/CuMeYL3Bcxb1prtUy3iLR13YAPjDFzgY+BSdba1z3u0/9v745pAAAAEIb5d82HibUyFhLgXMQDAAAABFgCAQAAAASIQAAAAAABIhAAAABAgAgEAAAAECACAQAAAASIQAAAAAABIhAAAABAgAgEAAAAEDCl/ZslSuWpNgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1440x144 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#convert the Pwm dictionary to an array and plot the PWM sequence logo \n",
    "import numpy as np \n",
    "pwm=np.asarray([m.pwm['A'],m.pwm['C'],m.pwm['G'],m.pwm['T']])\n",
    "print(pwm)\n",
    "plot_weights(pwm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Question:\n",
    "Why do the values of the rows in every column in the PWM sum to 1?\n",
    "\n",
    "##### YOUR ANSWERE HERE: #####\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Accounting for genomic background frequencies of nucleotides i.e. converting a PWM into a position specific scoring matrix (PSSM)<a name='PSSM' />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A PWM records the positional nucleotide probabilities based on an observed sample of binding instances (subsequences) from the genome. \n",
    "\n",
    "#### But what would these probabilities look like we just randomly sampled subsequences of the same length from the rest of the genome?\n",
    "\n",
    "Imagine a genome that is very rich in Gs and Cs. Then if I randomly sampled subsequences from the genome, I would get high probabilities of Gs and Cs just by chance.\n",
    "\n",
    "Hence, we need to <b>normalize</b> our PWM by these <b>random background probabilities</b> of nucleotides from the genome.\n",
    "\n",
    "For simplicity, we assume that the bases A,C,G,T all occur with equal frequency (p=0.25). Note: this assumption rarely holds in reality! What you would actually do is count the total number of As, Cs, Gs and Ts in a particular genome and then divide by the size of the genome to obtain the background probabilities of each nucleotide."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': 0.25, 'C': 0.25, 'G': 0.25, 'T': 0.25}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m.background"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The normalized PWM, which we will call a <b>Position-specific scoring matrix</b> can be represented a log-odds matrix i.e. at each position 'i' for each nucleotide, we record the log2 of ratio of the probability in the PWM divided by the probability in the background.\n",
    "\n",
    "We calculate the PSSM in accorance with the following formula:![PSSM formula](../Images/11_pssm_formula.png)\n",
    "\n",
    "#### Unlike the PWM which has all values between 0 and 1 (probabilities), the PSSM can have positive or negative values in each position.\n",
    "\n",
    "Positive values indicate a positive log odds i.e. greater chance of observing that nucleotide relative to background.\n",
    "\n",
    "Negative values indicate a negative log odds i.e. lower chance of observing that nucleotide relative to background."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': [-inf, -inf, 2.0, -inf, 2.0],\n",
       " 'C': [1.0, -inf, -inf, -inf, -inf],\n",
       " 'G': [1.0, 2.0, -inf, -inf, -inf],\n",
       " 'T': [-inf, -inf, -inf, 2.0, -inf]}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m.pssm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We want to avoid the -inf values in our matrix, as mathematical operations break down for infinite values. To avoid this problem, we can add \"pseudocounts\" -- small numbers we add for each position in the PWM to avoid taking the log of 0. We add the pseudocounts to our \"counts\" matrix and recompute the PWM & PSSM. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original pseudocounts:\n",
      "{'A': 0.0, 'C': 0.0, 'G': 0.0, 'T': 0.0}\n",
      "Updated pseudocounts:\n",
      "{'A': 0.001, 'C': 0.001, 'G': 0.001, 'T': 0.001}\n"
     ]
    }
   ],
   "source": [
    "#get the original pseudocounts, they should be 0 \n",
    "print(\"Original pseudocounts:\")\n",
    "print(m.pseudocounts)\n",
    "\n",
    "#Update the pseudocounts to 0.001 \n",
    "m.pseudocounts['A']=0.001\n",
    "m.pseudocounts['C']=0.001\n",
    "m.pseudocounts['G']=0.001\n",
    "m.pseudocounts['T']=0.001\n",
    "\n",
    "print(\"Updated pseudocounts:\")\n",
    "print(m.pseudocounts)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'A': [-10, -10, 2, -10, 2],\n",
       " 'C': [1, -10, -10, -10, -10],\n",
       " 'G': [1, 2, -10, -10, -10],\n",
       " 'T': [-10, -10, -10, 2, -10]}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Re-calculate the PSSM, we should no longer observe values if -inf \n",
    "pssm=m.pssm \n",
    "for base in ['A','C','G','T']: \n",
    "    pssm[base]=[round(val) for val in pssm[base]]\n",
    "pssm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[-10 -10   2 -10   2]\n",
      " [  1 -10 -10 -10 -10]\n",
      " [  1   2 -10 -10 -10]\n",
      " [-10 -10 -10   2 -10]]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABIYAAACMCAYAAAAJO6J1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nOy9eZQcZ3nv/6mq3rtnkTSj0W5ZlixLsuQF27IlE64Bg+2AjYw3Wc4lF34hCeQCCdg+yT3nXu7JL+f+sIMvhIQQkpDgfd8B25jNIHmTsLXL2vdtRrP29FrL74+na7pn1KPRMtPdmnk+57znra6q7qrurn66nu/7PM9reJ6HoiiKoiiKoiiKoiiKMvYwq30CiqIoiqIoiqIoiqIoSnVQYUhRFEVRFEVRFEVRFGWMosKQoiiKoiiKoiiKoijKGEWFIUVRFEVRFEVRFEVRlDGKCkOKoiiKoiiKoiiKoihjFBWGFEVRFEVRFEVRFEVRxiiBap9AKU1NTd7MmTOrfRqKoiiKoiiKoiiKoiijhjVr1rR5ntdcbltNCUMzZ85k9erV1T4NRVEURVEURVEURVGUUYNhGHsG26apZIqiKIqiKIqiKIqiKGMUFYYURVEURVEURVEURVHGKCoMKYqiKIqiKIqiKIqijFFUGFIURVEURVEURVEURRmjqDCkKIqiKIqiKIqiKIoyRlFhSFEURVEURVEURVEUZYyiwpCiKIqiKIqiKIqiKMoYRYUhRVEURVEURVEURVGUMYoKQ4qiKIqiKIqiKIqiKGMUFYYURVEURVEURVEURVHGKCoMKYqiKIqiKIqiKIqijFFUGFIURVEURVEURVEURRmjBKp9Aopy1uN54NngZKV3bfAcWfZ7d8Bjzyns569zC62wjHv8egbsc6rr8cB1wfDAo3BsT9Z7gGHI++nrTaCwjFFYNGS91+8DGNAbhWVPlj2v8JpeyW79XqDchzr4LsZxC4Pv6O/Sd2r+eRgl50nJ49KneyWfT+l79M/NLb4X/z0aBmCCYRWWC+sMs7De31Z4bFrl1/vLp7zeAiMgr2sESh6XLBtW/8dmEMxw4Vz8t1m4pt08uLlCnwcvD05OejdfvLb7Xa/l+pPYh5J9T7SPf7zS31m/35JT/F69gdemv84rs47+68tepyW/g9LrxTD67+N/7/2eV/rYPP57MCz5Lvp9l4X+uO/Y709mn1PY1wyCESxcE6FCX7JsBPtfJ6MYz/NI5pJ0Zjr7ta5sF52ZTjrSHRxKHmJnx04uarmI8dHxNEYay7aGSAORQKTab0lRFGVIXM+lJ9tznO3z7V9HuoO9XXs50nuERS2L+mxfQ7ihrO0LWaFqvyVFUU4SFYYUBaBjLaz5S8ApOMW+M2qDkyk4x1lxit0Sx9h3SKHEUTeKfT+H0H/sLxeIToPLvivPsaLifPVz5kod+1KHv3SfgU6eLxgYxd4Xh0od4+OWKXnsL5f03kBHeSBDba9FhhCY+n1/hb7sd2qW7F/63Zf5/D0XudYK4l0gXvjOyuA6YCch3w12j/T5nuMf20mwU+CkwEmDnQY3I8tOtnAdZ+X67Xcd+6JKqUjpFM5/gIgVbJBzNSNghQt9qCAshQrrwkVBwQgWrt3CNWwGC9dvoCg4WIXnGqHCawUGiBIlr+M//7jvxSj53BnwuGS9MXD/wR4z4PV8Sn8z/qpyvyd/X4rbjnu+V+b5Jfv32Zd8UZgrFejcHDi+cNdbsE/+NqdEvHNKRDy7sH+pPcvC0scgPL789XeWMphzUyrstKZaae1t5Vj6GO3pdrqyXSSzSZL5JJl8Bsu0CFpBABzXwfEcXM/F9dx+x3p1x6sAGBgYhoFlWJiGScAM4HouWSeLaZjEgjESwQT14XoaI42Mi46jKdZES7zlhMJSY6SRSCCCYQy8HkcBkybBkSPVPgtlLNLSAocPV/sshh3HdejOdpcVdvx2tPcorSmxfR3pDrqyXfRke+jN95K1swStIEEziOu5Yvdc6b0B93g/2fYTQGyfaZiYhollWliGheM5ZO0sATNAPBgnHor3iUfjo+NpijUxMT6RCbEJgwpLjZFGwoFwNT5GRRmTjLgwZBjGdcB3AQv4N8/z/r+RPqainDI9W+Hor05yZ7Po7FqRogM5mMPnO/99TqFb7AHq5sCkj53eeQ+MFPIjKfqOQfF8+lEu2mGAU90ndgwQPo6LfhjkdQeuM05yP4DnJkNmLDkLBiz8X9C9BXp2QPqgiD6+KGmG+guCgFxPhcgzN0/f9XSqNH8YJn4Ywk0QbhaBINgIoQYI1IsQFIjJOfgCw3Ei44D3UirM9IljvlBplSwrw05f1KAfeTjgu/IflxJqrPx5DsGJnBt/1How5yaVT5GxMwTNIEEriGVaGBh4eDiuQ97Jk3Nzxx3TNEyCZlAcmVC8T9QxMKgP1/c5M5MTk5neMJ3Jickk80kCRoD9Pfs52H2Qw72HaUu10ZnpJJlLEjSDxINxAPJOntZ8K4eSh45zsEzDJGSFCJpBzMJv3PVcbNcm7+bxPI9oMEoilKA+VE9DpKHvfJpiTTTHmk8YsRQPxkensKQoowzbtenKdA0q6rSn20XUTrXSnmqnI9NBd7abnlwPvbleck6OkBUiYAaOs305J0fezR93TMuwCJgBAmaARCiB4zlk7AwBM0BjWETsifGJTEpMYkbDDJpjzaTyKTw89nXt42DyIEeSRziWPkZnppPeXC/hQJhEKIGHR9bJ0tvby4GeA2WP7Z9vqe3Lu3lyTq6sqD4+Op4JsQm0xFuYEJswqKg0qkV1RRkBDG/ICIAzeHHDsICtwLXAfuBdYLnneZvK7X/ZZZd5q1evHrHzUZQT0dsLBw9CJgPZbKHPOOQyWexsFjubwc5mcfIZnFwWJ5/FzWewc1mcXAY3n8VzbTxXnHWvTPqY4ffYGDgY5LGwMQybUNAlHrWJRhwCloNlOgQCHgELrABYliwHAmCaYAUMLBNMy8CyDCxLlk3TxLTANI2+7CbTNDANMEwPwzQxfZ/dNETuMSUqxDBMZKO/bA5YNjAMq0TDKfNnW/YP2Djx5gFClbnpbzHs5Gl8i6OB0hS3CnDngGP50RCDRTCNBZ6dNIaESRPudIb9VX3n5v3D7/PugXc5mjpKW6qNYylxHDqznXRlukjmkmSdLHknj+MN/3kox2NgEDADBK0g0UCUunAdjWERkMZFxjEhNoGJsYm0JFq4bvZ1TK6bTF2oTp2rkaAWPtMR9APGIjknR2emkzf3vcmGoxs42iu2rz3d3k/c9qNzbNdW21chfNsXskLEgrE+21cawdkUa2JSfBKfueAzjIuOIxaMqe1TRhWGYazxPO+ycttGOmLoCmC753k7CyfyOHATUFYYUpRq8e1vwze+UW6LBcQKrYj/HzEc91PXXQc/+9npPdfzyp9DX7mbs5UL76n2GYwoA783M3sIerZB717o3Q092yF7DPJd0uwk5JOFtLC0iI19UUTmABGnNDKttM5USe2dUp6IwfhLJU0s2ADBeulD4wpRRBMgNB5C9RBogEB0QC2aQDG9q7RGUF/U2sA0xXIMTN8qScfsq6nk18dRhpfykWY5J3fCUetj6WO0plrF4SkZtU7mkvTme8k7eUJWiLyTxz3daLYzxBiyDtnZwcDoouF4vbybJ+/mSeVTHEsfG3Tf8OthSSdxHSLBCPFgnLpwHQ3hBsZFxzEhOoHmWDPN8WbGRcYdF6nkL9eH6/uiARSllsnYmaKAU8YGtqXbaO1t7Sf2+LYvlU/heA4hK0TGzlTtPajtG/z1fNvXm++lNdU66L73/Pwe8l4e13OJBqLEQ3Hqw/U0hBv6IpaaY81DRmuqqK6cTYy0MDQV2FfyeD+weISPqSinzKWfWs3iwDfJZD1cFxxX6jS7hVq3tgNOoblOYXtJ7+/rO/yeX9+5pHRIqY/cV7MYeM2E8V8KMaX9DhaYy5g8MUQkIpFBwaC0UKjY/Mf+Nr+V7m8VIossSwJ9Spu/zjBkWSKKim2wx8Xoo+NFJ8Mov1zu8emsM4yitlBOCDuZdeX2SeczPL35CV7Y+iy2Xyuq7wn9n1daZ9p/XKp3lG4/bt8y692+ckMGMyZHSdmp40+QukIr8+bskqLLrk2/mkX9Uon8ull+7SBfvCnuvyh5kD+f5DAj2FWsQePXH3Jz9Cvw3K8mUWlqklGoGeTXHPKLWocKLVBSS8iS3he3rNJ6RP6+/mO/zlBpjSL/IhxYX6hMGltpvS+gr+ZT2RpDBcf1kgdKXsPsnybat+yWXCClP27/oigYhOOKmBc/s/c79vP9bb/lQLqz3zVHz9bCdzXyTDThwEOf6EsB6Mp00ZPrIeccn2o1GAYGlmn11ZgImkHCltSFOJnCo6Zhyg27kyfrZAlb4b5UqZZ4C1PqpjCtfhqTEpP6RIiJ8Yk0x5ppijWNWA0K24aeHkgmIZ2WSNJcbvA22PZsVp7vt0wG8oVsDv+SOe7/o7Cu7zMquWyLzcPxbHJulryXwTXTuIFebCNJ2mgn5baT8jrodTtJu11kvG4IZAkGXayAA1Yez8qClQMrh2HlIZAtLoeSGOEkBNMQ8HA9ty+9b4+355SiHAyMvhF637FqjDRimdZxNZsqhYHBh2d8mC9c+gWaYk1VOYexiOd5vLHnDX74+x/Smems2nlMSUxhZ+dO2lPtfVE8yVyybKrVYPg2zzTMvpQoXwANhoIn9fzStKloIEpDpIEJ0Qm0JMT2Ta+fTku8hea4iBAT4xNpjjczITqhrwbacJPLid3r6RF7dSL7diL7l8lIK7V9tl3e7sGp2T7by4vtczN4VhrHSmJbPaRpp9dtJ+V2kHK7SHti+8xAnkBIbJ9nZcEUu0egYO+sHFhZjECuz/ZZgSwYYvu6MpLC7NeaO1kMjD5RqTHSyLjIOGaNm8WDyx4c5m9NUc6ckRaGykmk/dwzwzC+CHwRYMaMGSN8OopSnmlNDcyZMY4XP3iR7mx3X042IL6iCYzM/y8u0AHY03/BPuNP+eIlf8YXLvkyU+LT+wQnp0R8OtXlXO7E+zgl9/YDxZdSMeVk1g3n/gO3DZw0rXT5ZLaVLh/O7eAnbf/AL9p/BEDGrXTamqSL1QfHce30ZZw3roWXtr7AprZN/a+9CvF6V4jvHD7E0ulLuWfp/+Tjsz5++qP7feLHgFnJBp2FrMwMZkPuM2CWsoH7yokMOCcGrC+3vdwFWSjEDdKbJbW2fI6blWyw7QYZO8eTu1Zy//oX2Jk8QsbJV8UxtgyLxdMWc92CO/hEFZ1i0zAldalE6DkTZ8d1obMTWluhvV0cG7/5jk5nJ3R0SOvqgu5uWd/bC6lU0YFxnKLo7gvlcLyIPVDMKW2OM+DyG1YM5I8pCCTK7lE6GBA2wTQLYiUuHh5G4Zr35U15PwaeC65r4NgmrmsQDLqEwg6xqE04YhOJ2UTiNtFEjnh9nnhdjnhDjmg8TyRmE43ZRArLsbocdY1Z6sZlCYWL13rOybG5bTOPb3icPV17qmL7frnrl3zzN9/kU3M+xdeXfJ3FUxdXdnS/paW6xbdbWip2qK5MFz9e+2O+verbtGfaSeaqky4eskJcM/MaFk9bzEfP/WhVzgEgYAb6CjA3x5sZHx1PwDx9t8xxxOb5ts+3dwNtX3u79L7tSybF9vX2FoUc15UBSMsqDiCW+1kMZvv8gdSRwwBChVZm8Iz+A6MR0ytUQyjaPekL76Pv/Rh4roHrGDi2gYdBMOgQjDjEwg6RqNi1aNwmmijYvXppkbhNNJYnEi/Yx1ieeH2eunEZ6hpzBIJF2xcNRkfuo1GUM2CkawxdBXzT87xPFh7/NYDnef+n3P5aY0ipNq7n8vb+t3l609M8sfEJ2tPtfQVIK4U/0n71jKu5Z+k9Z+akK304rsNPt/2Ub638FmsOrZEitKcwMnim+NNVT4pP4vYLb+fW+bdy6eRL+zkhh3oO8dLWl3ho3UO8e+BdQlaInlxPxc4RIBFKkAgl+Ksr/4ovXPoFxkdH12xV1WJnx06+9/b3+Lf3/g2gok6RgdFXUPTaWdeyYuEKrpt9HXXh8jfUtUQqBW1t4uy0tRWXjxyBAwfg0CE4erTo7KRSxchJf6TZ88RJsW2J1HErr8ONGUqjVn1n0g9wzGRkW10djBsHTU2wapU8b2fHTp7f8jwPrXuITa2bCFmhiv5GTMMkGogyuW4y9y69l+UXLiceilfs+KOZtYfX8sBbD/DkxicxDZNUvlx07MhgGVZfjZib5t7EHRfewUfP/Wjf/3Gt4nki2JSzfYcPF21fa6sI3J2dImiHw0XbB0WhRm3fyGIYxUEEX5AC+R7zeYmmCoWgvh7Gj4fmZnjjjeqeszJ2OVGNoZEWhgJI8emPAQeQ4tN3ep63sdz+KgwptcYHbR/w3JbneHjdw2xv307ADNCb763Y8ROhBHWhOv7qqr/i85d8Xp300+Bo71F+uOaHfPft75K1sxUVWhKhBDknx4LmBaxYuIJl85Yxa9ysk3puT7aHV3e8yqPrH+XVHa8SMAIk88mKRZdEA1E8PG6ceyPfuOobXD718oocdzThuA4/2/4z7lt5H+8efLeiYmTADBC2wkQCEZbNW8YdC+7gD875gxFLPThVXFccnP37Yd8+6XfuhG3bYM8eEX66u2W/SKQoNDhOMU1BOfspdwvalmrj5a0v8/C6h/nd3t8RDoTpznZX7JwSwQQuLv910X/lq1d+lQuaLqjYsUcLWTvLU5ue4r6V97GjYwdZO1uxAsthK4xpmIyPjuf2Bbdz64JbuWLqFTUzwGbbMtGJb/v27YMdO6Tt3StCd0+P2Ds/ase3fdlsMQ1VObvRmu9KtaiaMFQ4+A3Ad5Aqvj/yPO/vBttXhSGlljmSPCLRHGsf4u0Db1c0miMWjOF6rjrpJ4nneazct5L7V97PazteA4OKFYKsC9WRc3J8+JwPc9fCu/jD8//wjOtX5J08v937W57c+CRPb3qajJ0ZdNrZ4cYfSZ9aP5V7l97LHRfeQSwYG/qJY5ijvUf51zX/ynfe/k5FxUh/FHxyYjLLL1zOLfNv4eJJF1et8OWxYyL07NolbdMm+OADcYSOHStGl0CxjkUtYZoQi0EiIS0eH7yvq4OGBmhslFHhhobitlhMHLzSGm2D9QOX/Yib0loc/vJg6/yWTksElZ9K0tUlrbOzfwpJub50eSS/l6FuQVP5FK/vfJ3HNzzOy1tf7ltXCZEhYAYImkEunHgh9y69lxvn3lgzwmqtsrtzN99753v865p/xcOrWNRXPBjHdm3mjJ/DikUrWHbBMuY2za3IsQfieSJ6b98Ou3eL4LNpk9jCffvk9+dH9rhu/5pjtYJlndjelfaJhEQANjRIq6uTFovJPoHAyNm+0vqefu84Yvt8++WnCnd2npztK+1H8ntRYUipFlUVhk4FFYaUs4VkLslrO17j0fWP8sr2VzANk95874hHc5iGSSQQYXr9dO5dei+3X3i7Oukl9GR7eGjdQ9y/6n5ae1tJ5VMjXrPCNEziQUk5+NT5n2L5hcv5+KyPj1gOued5rD2ylmc2PcOjGx7lYM9BDAzSdnpEjldKIpjAw+NzF32Or175Vc6fcP6IH/NswfM8Vu1bxf2r7ufVHa8ClREj/ai0hRMXcteiu/jMBZ9hZuPMET9uKbkcbNkC69fDmjXw1luwebOIEpGI3ACnUiNdc+LEhEKSutTUJGH8ft/cDNOmwZQpMHEiTJgg4k40Kg6NbUvza7L5t0y+E+MX+g8NXWO7pnAccXr6JlUYUOzVNOV9+U5dNivOlV/D5PBhiXo4eLB8uktX18mdx6ncgjquw5v73+SpTU/x5MYn6c5247hORVK960J1WKbFX1z+F/z55X/OlLopI37MswXHdXh1x6vct/I+3j7wdsUiI/1BmMXTFnPXwru4ce6NtCQqVzMJxK5t2gTr1sG778I778DWrfLbikTkt5VKVTeFKxo93vY1NUl5qalTi7Zv/HgRdqJRsWv5fNH2lRaI9icu8W1f8CzTSn2bbttFYcmn1PYFg/Kes1mxZx0dEs3lR7oePtzf5vl98iS10Bpyv5UxhgpDijKC2K7N7/b+jic2PMEzm58hlU/1zTAxkiRCCTzP448v/mO+svgrY9pJ33B0A//3zf/LYxse6xPpRpKQFSJgBmgIN3Dbgtu4df6tXDntSiyz8tOp7+ncw/Nbnufh9Q+z7si6itTmCJpBAmaARS2LuHfpvXx67qfPqGjm2UwylxQxcuX9HO09WhExsj5cT9bO8pGZH+mLSqtUmumRI/Dee7B2Lbz5Jrz/vtS7iEaLdTEqjWGIc3PuuTBzJpxzDsyZA7NmiegzcaI4aaUz4pimiDnhcPmiqsrJ4dfP8J1Hyyqmv3R0iHC0Z08xcmz3bmm7donDfLq3oJ7nsbltM89tfo5H1j/Czo6dWKY14vVr/BqA18y8hruX3s01M68Zs1NRt6Xa+tK00/n0iEdG+oMwrudy/ezrWb5wOZ847xMkQuULrw83e/aIvXv/famNtX69CAGxmFz/vZWrMtCHZcH06WL3zj0XZsyA88+X5WnTRAAKBovF9Et/o+GRmcxxzODPxOYPeFhW8TM9dkxs3+7dYvtK7d6ePfK8GnK/lTGGCkOKUiE8z2P90fV90Rz7u/ePeDSHZVhYpsXs8bNZsXAFS6YvwTIqL1BUmryb5409b/Do+kfZ07UH27FxGblhuVgwhuM6zBo3ixULV3DzvJu5oOmCmnIK2tPt/HTbT3lk/SP8evevJd0x2zOiQkU0ECVgBrh53s18Zu5nmBCbMGLHqiV2deziyU1P8vOdP8fEJOOMXHRQaQHVG8+/kTsuvIOPzfpYRQqoptPw29/Cyy/Diy/KKGk4LOsrnf4QCIjTs2gRXHIJXHUVzJ4tzk8+LzfqllWM9lFqi3RaHCLTlO8ok5GUk+HgYM9BXvqgULj/4LuErfCICxVRK0oinOCOC++QYu6h2i/mfqZ4nsfG1o08vuFx3jrwFsCIDoIFzSAhK0Q8GOez8z/LbQtu4+oZV1dkIKKrC371K7F7P/mJRMwFgyIAVTr6MRyGefPE9l12GSxeLOLPuHFFwTsQEJHKrI1SSkoJqVT//6eeHvnuFKUaqDCkKFVib9deXtjyAt956zvs7NxZseOGrTBhK1xTosVw4XouOSdX0ZniLphwAV+78mvcOPdGJtdNrthxz4SMneGXu37Jo+sf5dH1j1ZsGmi/6PFojSDKO3myTmULqX7p8i9x6/xbWTxtcUUKqLa2wiOPwBNPwO9/L9E2yWRl0yFMUxygq68WAeiSS2RkPFPQ3xIJdYCU8vRke3hl+yv82+//jdd2vlax4wbNIOFAeFQOzHh4Yvvs7IgOwJTSEG7ga1d+jZvn3czCiQsrcj+zezc8+CA89ZSkhIXD4sRXkmAQli6FJUvE9l10EUyaJOKCaQ6fmKooythEhSFFqQH2dO7hvlX38dTGp2hNtVb7dJQh8Av4fv2qrzOl/uyuJ2G7Nv/5/n/yg9U/4L3D71VsZjPl9AhbYT5yzkf46pVf5YY5N1TsuL/+Ndx3H/zyl+KApEe+bFU/xo+HT34SbrkFrr1WQu3DYU15UE6fVC7FP737T/z7e//O1mNbKyaQK6dHPBjn+tnX840l32DxtMUVOabrwvPPw9//vaTJum7lC+FPmQLXXw+33y5ieC4nkSVnW+0yRVFqHxWGFKXG6M508w/v/AP/+f5/sqNjR7VPRwEMDOZOmMvnL/k8X778y8RCo7eo9yvbX+G7b32XX+3+VUUjr5TBqQ/Xc+P5N3L30rtZ1LKoosfevRv+5E+kbkZqZEu0lOXKK+E734GLL5ZUo/r6yp+DMvpxXZfHNz7OP73zT7xz8B1s1672KSlAc6yZ2xbcxt1L7uacxnMqeuz334fPfU5mDqtGjaAbboAHHpC6aLYtkZCKoigjiQpDilLD2K7NQ2sf4p9X/zO/P/T7EU9RMTAqko5yprieO+KjuwEzwOKpi/ny5V/m9gW3Y47B3JQNRzZw/6r7eeGDF+jKnuRUQmeAiVnzKY6VuPYAptRN4c4L7+TrS77OpMSkET9eOXbulDStatTNiEbhySfhmmtkWmNFqSS/3fNbHnjzAV7b8Rope+QVUdMwMVDbZ2Bw3rjz+OOL/5j/fsV/pz5SHSX4zTfh4x+vjhje0iK1i+bPVzFIUZTKosKQopxFvL7zdf7pnX/irQNv0drbimmYwzr1q4GBZVo0RZv4yMyPsHDiwpqoB5Nzc6w9vJbf7PkNXZmuYZ/uNmSFcFyHSYlJLJm+hK8s/gpXz7h6WI9xtnMkeYQH3nyAn277KR8c+wDDMMg7+WF1FEKWxMZfMeUKlkxfUrHZtIZif/d+3tjzBpvaNgEMazSBaZhYhkXICnHhxAu5bcFt/Pllf040GB22Y5wul18udYSqMZ3yj38Mt94qApGiVJMd7Tu4f9X9/HLXL9nVsQvLtIY9mtKfzXLp9KUsnrq4YrNpnQgPj+3t23ljzxvs7JA6iMM5OGUZFoZhkAgluLjlYj530ee466K7qn7P4XkydfuxY9U5/ltvwaWXnn1TvSuKcvajwpCinKW09rby8taXeWjdQ6zat4pwIEx3tnvYXj8RSmBg8MUPfZG/uOIvmNk4c9he+2TZemwr333ru/x47Y8BhnWq+fpwPTk7J9N6L7qLG+bcUDNCRK1juzYr967kyY1P8vSmp+nN95J388M2A03ICmEaJounLuaepffwyfM+iWVWtmhrOp/msQ2Pcd/K+9jXvY+MnRm2+kvRQBQPj2n101h+4XI+O++zLGpZVHPRUlOmwKFD1Tn2u+9KgWlFqSUydoZf7PwFj298nBc/eBHP80jlU8MmmPizCX7yvE9y95K7WTJ9ScXtQke6gx+9/yMeePMBurPdJHPJYXvteDCO7dpc0HQBKxauYNm8ZcweP3vYXn84cBypXVbpKEmftjaYMDYm8FQUpcZQYUhRRgG9uV5+vvPnPLbhMX667acYGPTme4fFkfWd9CunXsndS+8ecSfddm1e+uAl7lt1H2sPryXv5oclQmPgtN7LFy7no+d+tCLTeo9mPM9jw9ENPLP5GYi0RzcAACAASURBVB5d/yj7uvdhGiap/PDE4NeF6ogGo3x18Vf5k0v/hOZ487C87mBsO7aN7779Xf7z/f/EwCCZHx6nqC5UR9bJclHLRfzRoj/ipgtuYkbDjGF57ZHihz+Ev/zL6qRTXHaZTAcdCmmRVaU2cT2Xt/e/zdObn+aJDU/Qnm7H9dxhiSYyMIgFYzTHm7lnyT3ctegu6sIjO+XUmoNreODNB3h2y7MYGKTt4akwXxeqI+fkWDpjKSsWruDT5396xO34mfJXfwX/8i/VsX3LlsHDD4s4ZY2+SewURalhVBhSlFGG7dqs2reKpzY+xVObniKZSw5bNEcilCAWjPG1xV/jTz70JzTFmobhjIVDPYf4weof8I/v/iN5J09P7szngQ1bYUzDZHx0PLcvuJ1b5t9SsWm9xyr7uvbx4gcv8tC6h3jv8HuErfCwfJd+lM0Ns2/gG0u+wZXTrhy2kXTbtXl568vct/I+3jv8HrZrn7EYaWBQFxaH6JqZ13DXoru4fvb1jIuOG5ZzrhTf+hb87/8tM+FUegR96lSZDejGG2UmtIhquEoNs6VtC89tfo5H1j/C9vbtBMzAsES5xoNxXM/lzoV38pdX/iULJi4YhrMV0vk0T2x8gvtW3seerj3DEhlpGibxYFzs9ZwbuPPCO7n2vGuJBc+eSRtcF778ZUlprfQMjADz5sH3vifF94NBFccVRakMKgwpyijG8zw2tm7k2c3P8sj6R9jTuQfLtM44mmO4nHTP8/jNnt9w/8r7+cWuXwCc8WirH6o+Z/wcVixawbILljG3ae4ZvaZyenRmOvnZtp/xyPpH+MWuXxCyQvRke86oLpE/kt6SaOGeJfewYtGK067HcTh5mB+s/gHfe+d7wyJGBowAkWAEy7C46YKbWH7hcq6ZeQ3hwNk9p/ru3fCVr8Brr4FhQCZT2eOffz7cfDPcdpsUZNXZyZRa53DyMC998BIPr3uYtw+8PSyp3n4tsguaLuDepfeybN6yvrpsp8qO9h189+3v8h/v/QcYnHG6mF8jqS5Ux63zb+W2BbexZPqSiqcADzdr18KXvlSstVbpqeovvVRs3y23wMyZcvy6kQ0cUxRlDKPCkKKMIQ50H+DFD17kwXUP8vtDvz/jaA7TMIkGokxKTOKepfewYuEK4qGhpxDqynTx47U/5u9X/T0dmY4zvin1Q9UXT1vMXQvv4sa5N9KSaDmj11SGl6yd5de7f80TG5/guS3PYbs2GTtzRpE5/qj0XQvv4mtXfo15zfOGfI7nebyx5w3+ftXf8/OdP8cwDDL26SsdkUAEA4OmWBO3X3g7t86/lcumXDYqo9KOHIEf/Qi+/30pzGoYlU+1aGiQ2YJuuQU+8hFoapJzsCydwUepTZK5JK9uf5XHNjzGK9tfwTItkrnkGUXm1IXqMA2TP7vsz/jy5V9mesP0IZ/juA4/2fYTvrXyWzLLqeuc0UQOsWAMx3WY2TiTOxfeyWfnfZb5zfNrrlbacLBzp6SW/fu/izDuOJUXyCdOhOuuE4F88WIRx1MpiSaKnT3BWIqi1DAqDCnKGKUr08Ur21/hkfWP8PrO1wlawTO6We1z0hfdxdcWl3fS1x5ey7ff/DZPbXrqjOrQ+KHqrudy/ezrWb5wOZ847xM1MZOLMjSu57Lm4Bqe3vQ0j298nNbeVjy80xZoAkaAoBVkfvN87l16L5+54DMErf5TunRnu/nx+yJGtmfaz0iM9KPS5k6Y2xeVNmfCnNN+vbOR3bvh1VfhmWfgd7+DQEBSLuzhm7DtpIjFYMECWLhQZlG74gqJMLIsiSwKBGQfc/TpdKOWbFac7oaGap/J8GO7Nr/d81ue3Pgkz2x+hlQ+dUap3mFLohGvnnE19yy9h4/P+vhxovSR5BH+Zc2/8A9v/wM5J3dGg0F+rbQPTf4Qdy26i5vm3sTU+qmn/XpnG54HmzeL7Xv6aVi9WlJck8nKz97Y0AAXXgiLFolQ9KEPwezZIlrl85KCFh96nE6pIdJpsX+NjdU+E2WsosKQoijknBy/2f0bHt/4OM9tfo68mydrZ09rNDFgBgiaQRY0L+Deq+/luvOu4/kPnue+lfexo2MHWTt7WjO4hKwQQTNILBjjlvm3cNuC27h6xtVVn9pWOXO2t2/n+S3P8/C6h9nStoWgGTztos/+SPqXLv8SX7r8S7Sn20WM3PgUhmGcthhZH64na2dZMn2JFFCd+2kmxiee1muNNmwb3nlHnKWVK2HDBmhvl6nmbbs6BVynTBGB6NxzYdYsEY9mz4bJk+WmO5eTZhhS5FXrF1WGfF4cH9sWsS4alXVHj8KePbBli7Tdu+G556p9tiOL53msO7KOZzc/y6PrH2V/z35MTFL26f1gEqEEiVCCr1/1df7bxf+NTa2buH/V/fx8x88ByDinLrwbGNSF6si7ea6ddS13LryT6+dcT31YczlBBMyVK+GVV+DNN2HTJujtFXuSy1U+qsgw4JxzYM4cST2bM0fqFZ13HkyaJFFGvoBvmmL7wmd3pvNZQy4nts9xZOAiGpXv4uhR2LVLBMetW2X55ZerfbbKWEWFIUVR+uF5HmsOreGZTc/w+IbHOdJ75LSjOeLBOBk7QzQYPa0IjdJpve+88E4+O/+zLJy4cFSGqitCa28rL299mYfXP8zKvSsJW2G6c6demyNshbFdm4AZwHbtUxYjBxZQXX7hcq6dde1JpUoq0NMDGzfCunUiGq1eLTe9ICPZqVTlo4t8TBNaWmD6dJg2TYpcT5wo/ZQp0NwsKWoNDZKeZtv9i29blrRAYOwWhfWjEmxbIiVcV5zSYFAczXweurqgowNaW+HwYTh4UNrRo7B/P+zbJ33vIPWZa+gWtCLs7drLC1te4KF1D7H2yNrTTvWOBWNk7SwhK3RaM4sFzSDhQJiQFWLZBcu4fcHtfGTmR067ntFYo70d1q8X2/f227BmjaSiBYNiN6oRXeQTCIg4Pn26tMmTxRZOnSrLzc0wYYLYvlhMfse5XPH3bZryHoJBaWMR2y42xxE7ZZpF25fJiO1rb4e2Njh0CA4ckP7IkaLd279fhKJyjDXbp9QOKgwpinJCdnbs7Ivm2Ni6kZAVOuOaQCfCD1W/eNLF3LXwrrNiWm9lZEjlU/x8x895bMNj/GTbTzCQiJ+TFXkswzolQciPSkuEEtw6/1ZuXXArS6Yv0ai0YcLz5GZ43Tp4/33pt26VdR0dMsoeCMgN92BiQTVobBShqKlJHKfGRknRiMelEGxDg6yrr5eWSEiLx2VUOBoVhyEUEkfCdyh8p2IoyungpesGvsbJvqbv5PmOnuOIE5hOi3PT2ystmRShr7tbHJ7OTllOJovb29qktbZK/anhKNJbQ7egFacj3cHPtv+MR9Y9wi93//KUC/ebhnlKaeF+rbSWeAvLFy7ns/M+y6WTL9VBmGHCdSUSZN06eO89EY62bxfBoLtbbIRlFX9/tYBpwrhx/W2fb9/icenHjZN1jY1iC+vqRFAaaPt8+5LPS++6J//7HngJDmb7TuX1fNsXCMi5+ZGMmYx8/r29MniRTMr3090tdq+rS5ZL7WKp7WtvH54Bj7Fs+5TqMmLCkGEYtwLfBOYBV3iet7pk218DXwAc4Cue57061OupMKQo1act1cZPtv6ER9Y/wht73ugb0TzTWab8ab0/OvOjrFi0ghvm3EBjRJOslSKO67Bq3yqe2vQUT258kp5cD7Zrn7A2x8kIQ34B1XMaz2HFwhXcPO9mFjQvUIeowuTzIhDt3i0O1O7dIhrt3i3r29rEmYhE5KbeFzEqPUvQmRKJ9HesEglxTExz8GZZg2/znazSVm7dwFYq+vjLzqln+I4o6hwJWTvLr3b/isc3PM4LW17A8RzSdvqEhftPRhhKhBLknBwLmhfwR4v+iJsuuIlZ42YN9+krQ5DNwt69YvcG2r6DB0Vo9dO+DKMYwZg//brhFccwRCQqtX3xuAgzp2Lvytm+k7F3pfuV2r1kUgSgakVwDYbaPqVajKQwNA9wgX8BvuELQ4ZhzAceA64ApgCvA+d73onv3lUYUpTaIp1P8/rO13l8w+O8tPUlgJOO5vCn9Q6YAT4z9zPcfuHto2Jab6UyeJ7HptZNPLflOR5e9zC7O3djmdZx9YMGE4b8qLRLJ1/aF5U2rX5apU5fOQ08T0Zq9+2TUfbWVhGLjhwR4ejwYUlROnZMRnVTKRFcQiFxIjxPHCq/xoNSW/gRVaXfVS1FjdUKruey+uBqntr0FE9seIK2VFvZVO/BhKH6UD1ZJ8uHz/kwf7Toj/jDOX/IhNiESp2+chp4nkSi7NsnQpEfnXL4cDFFyY9W6eqSiBc/Ssf/PeXzEg2jgkPtEYkc/11Voy6fokAFUskMw/g1/YWhvwbwPO//FB6/CnzT87w3T/Q6KgwpSu3iuA5v7X+Lpzc9zRMbn6Ar24XjOmSdYgK1H6reHG/m9gW3c8v8W0bttN5KZTnQfYCXtr7Eg2sfZM2hNX2RbL5z5Eel5Z08H5v1MVYsXMH1s6+nITIKpz1SABkB7ujoH+bv9wcPijPlp0Ylk8WRY39WGNeVm/VAQEaw/QAyz5NtpXUmxiqhUPHzKZ31zU+T89NGQiGJFojFpCUSxbS7CROkztPEicV0Fb+fplrtkGw7tk1Svdc/zAdtHxAwA/Tme/tsn2VYxIIyl/mnz/80yxcu52PnfoxoMFrlM1dGCtsu1rcZaPsOHBAx3U+J8iNnSm2f5w1u+wb+tscihlG0fX70Ehz/3+C6RdvnR0slEsX0Y9/Gldo8P21vypTqvkdl7FINYegfgbc8z3u48PjfgZ95nvf0iV5HhSFFOTvwPI8tbVt4bstzPLT2Iba1b2PBxAXctfAuls1bxuzxs6t9ispZhOfJzaqfz+/3pcvJpIyG5nLQncqy+dAu1h/Yxu62I1hujGmx2UwKzyRhNZHPmWSzxTQkPyTfb7YtvecVR1fLLZf+Pfo3zoYx+HJpPYPSFgpJi0Sk92eJiUT6N79Wgx9ZEYvJDaZ/o1lu2bIq8x2NRvL5wrXV49DblSTdkyTT00O2t4d8Komd7sHJ9uBkk7i5FJ6Tw3NymF4a00sTIINlZLDIEDAzBIwsQTNLwMwRMLMErRwBM0fQyhO0pA+YeQKWjWU6GHhggIGHYcjF1m/ZKF6AnmdIwyg+xgAPPGSb7QawnQB5J0TeCWK7QfJOCNsNFfowthsm70nveBFsL4JDBJsYrhHBsMIYVgjDCmOGE1iROgKROoLRBKF4HeFEHdFEglhDHfG6MHX1BrFY+RpJytB4nti1gfZuoO3z7Vl7Ty8bD+1gw4HtHOzoIOTVMT0+h5bIOcSMcWSzRj/b5xcWLrV9tj24vStd5zOU7fMdZ9/e+QXbS22fb9d8u1dq/6LR4rK/v1/bq5zt851vU8ebTpu+/9tum96uHjIF25dLJcmlerDTPbjZJHamB89OF2xfFsPNEiCFZaTF7hlZAkaGgJmVZTNH0MoSNHMErBxBM0/AyvfZwIBpY5ruSds+zxtg7yhv+8TeBbCdIHlH7F7eDWH7fcH22W4Y2wtjl9g+14jiGhE8M4IZCGFYEaxIAitchxVJEIzWEY7XEU4kiNbVEWuoI1Efoq5Orlm1fcrZxomEoSGrbRqG8Towqcym/+F53guDPa3MurIKlGEYXwS+CDBjhhafVZSqMWmSDDOdBAZSWGwe8Dd9a9cB9xTaKdLSIjHTylmJ6xZHL/2Ry9Iite3t/Ys6lo5gZjLS/NmO/BE6/2bLrx3gF/OVx2HggkIDG9hdaKMBPzqjNFKj9PPwixvn87KP71z5BUH90cr6eikcOmHC8aOV/nI8fpbf2HoeOCnItkGmVfpsG2RbIdMGuXbIdUC+E/LdkO8BOwl2iqCTZryTYbxngxkEIwBGYfg8YkDEA88Fz+nfRjUmGGbhcwgUlgseeMaDtANHHXALyqoVBisqLRCHYB0E6iDYAKFxEGqEcBNEmqUPN0G4sBwaD+bZrWzatti4cpEbx45JRFu5grbptLRcrvhbH8r2SSHzOLCo0CALbC+00YBfKN1v/ufhi1alM+UFg0Xb5xeKr68X+9fYKLZv/PjjozX8Pnq2B1R5Htg9g9i+VrF7uQ7IdxVtn9MLdoqwkybsZJjgOQNsnwlRxPbhgmsX7J6LlIwdzZjFz8AMACWhVCkPeh04bIvtMwwwC7YvEBPbF0hAsB6CjRBqEPsWngCRiWVsX2PRripKDTGkMOR53sdP43X3A9NLHk8DDg7y+j8EfggSMXQax1IUZTg4SVFo1B1bOQ7HkTScw4f7Ozt+rRe/3sGxY0Vnxx/pLa0fksmcWsHHsRq2PpDSUP6h8CMBeoaY8dr/fvwII78Wj/8aNYPniSOTOiAOju/sZI5C+gCkDkHmMOSOQa4g9oAIFEbhlsZzwM1JO1ncHHCWVbkeEdyCGGYjssMQOGlpJ8QEK1Jwtgx5fTcPbhburK2KsLmcpCEeOdLf9h0+XKxz5dd66e4WG+dHx5xurRdf/FGKtu9k8KOiurtPvJ9fi8eyiuJSJlODn7nnQfYYpA8WBR7f9qX2Q/oQZI4UbF+XiEKYYIVKbJ8NTg68UzDqavsK+LYPsU0nwkMGJJyUfB+DYVgiIPXZvoKo7ubhTr3hUWqPkZqf90XgUcMwHkCKT88B3hmhYymKoigniefJKLY/M8q+fbBjh0yru3eviD5dXcVpxQ3j5GaHymaLQkMt4c8QFYuVnx3lZJpfV+BUWy5XjI6qthNSM7N7OVno3Qu9uwr9HujZBsmdIvxkj8nNuRWREVVfpHAyDBJ4LFT7Ay6HESiMJMfFOfCjbwyTYmTOgOXjthWElD6nxT3+cdllB+xeaUMKNyONW3Ciymx6dhLcXJloUc+T4uW7dsGePWLvfNu3b5+IPslksVCsLyJksye+vPyox1rDjyCMxU5+Fqjhsn2ZTO3MhFcz/012Cnp3Q3I3pPZK37NNbGH6oNg+wwIzVLR9bh7cE11cDti1pOwXMIMSQROIgxEcOdvnOcdv9xyJjMonhxZ4RhrPqQnbpygnyxkJQ4ZhLAO+BzQDPzEM433P8z7ped5GwzCeBDYhUf5fHmpGMkVRFGX4aG2F9ethwwbYvFnanj0i/HieOD9QrNszkGrMmGEYkgLQ1CTpT349idKpb/00qcbGYnFbv/ZELFYsAunX6fHTriQNo/8o/mDLp3K+5R6X1t7wU0T8OkrZrKSQpFLFtJKeHhn19lNOurqOn263t1fW+xEM6Wr7+ifCyUL3FuhcJ33XJujZLlFAdreE3xuWOD3OIBdaNZwdK1IM9w82QDBRTBEolyoQbCisqwOrkE5gRQrRMQUhyM2LsNU3e1SZIi6l60+Jkguw38VoSDMsGak2LIkKcLIisDkpcVLtXvk+8j2FdJMuWPO10ziPMyAzvNGinicRP+vWwcaNsGlTcVrwo0dFIAmFZL90urzgUw3bZ5qS9uTbvoF2z6+rM26c2LzGxmLtHV8AGmj7/HpClbZ9fvPrrfniWiZTtH2pVLGWkl8gvqOjWFNpoP3r7CzavpoQuAfDTkHXxhLbtxl6dojo7aTEThhGwfaVMeKeXR0xIxAvpjv5ds0Xtf3lYIOkQAUbCnawrrjNikIgCmZEojehYPucCtq+kt60CoKUUWL70mCnC+l0yYLd65Go01wHvPf10ziPM2CYbZ+iDAfDUnx6uNDi04pSRapdaKSGbNHZRDYros/69bBmDbz1ljzOZkX8yWSqO1oaj8usHP6sRM3N0k+dKrNy+PUeGhpkX9uWG39/lNkwxGnyCzuHQtV7L5XCF7NK6ypBscB1JCLrfDHJT3U5eFBmpPHrnRw6JJEQhw6deNT+tH96nicOT8c66FwLbW9Bx/sy+m1FAU9uwKuFYUJkMsSmQXRK0fGJTZXH0UkFR6jg7BimOGWuTdFZMQtOhlWoxXF218QZkker8D9w5+ldgL29Iv6sWwfvvgvvvCMikOeJnUilqpsm2dBQtH3NzdJaWsT2TZ4sjydMkP2i0WKh6NLfu2/7/MLOo53SGkJ+9BEUhaZIRLZ3dxdnJPSndPeneW9rk+V9+yQlcDD7dka3HJ4LyV3QuV5sXtsqWc62ifjj2YOL3pXADIqNi00TG+jX+YoWbF+kpVjrJlgnz3GyhQicwgdTWmvMDI7+mjhnke1TlDPhjIpPK4qiKLVBPg9vvw2//S2sWgVr14rTH43KTW5ygA9eCUGosRFmzYJzz4WZM2HePJg7tygEBYMySuy64uT4s3MNhmXJiPdYxi+8OtTn4BeSnl0yCWBp/Qx/SuJwWBypI0ckjWbTJkmf8dMJT5revXD4F3DsLWh7B3q2Ap44DXaqUJemgD1E0aPhwAhAYibEz4X4TKibAw3zIHGuOEPB+pJaQ4ZEz/TVexiEE21TqkY6LXbvd7+DN9+USMi2NomWse3jo3wqEVHX3Cy2b+ZMsX/z5sH554vta2oSe+fXWfMjlU70m/ZF37GMb/uG2icSkf+XuXOL6/26dn6ElF9XraNDxKMdO2TQZPv2U7R7nie27sgvofVNaF8tqa9+RF6+l365QvYQRY+GAzMMiVlF+1c3BxrmQ/wcEX0CdeCmCyK3ITbaT88d9DXHgPKoKMoJ0TsgRVGUGmbnTnjlFXj2WVi5sii0lKZADFV4eDgIBuUmfNEiuPRSuPJKcYQSiaITFo2Wj+gZCyPdtYJlSeTVQMaPlzZvHtxwgzjSudwQTpjdC0d+AwdehAMvSw0M05L1pVSihk10MjQuhMaLoOkqGHexjIY7GRnlNoOS0jAQMwDERv78lGHF8yQa6NVX4emn4fe/FzEgmexf0H6owsPDQSQCCxaI7fvQh2DxYrGFwaAIEaYptq+cnRsLEY61QiAg/0cD8aO1Fi6EG28U2zdkSbJch4jg+56HQ68UbJx3vK2rRJGM+DnQuAjGXQQTroJxi0T8cdISuWSGJY1rIGaZD0NRFOUEqDCkKIpSY+zcCT/4ATz4oDg+hlEcEa9UkdNx4+ATn4Bly8QRmjq1KAAlEuIM+ajzc/YRi0k7DicDe5+Brd+DjvekZoTdQ196QSUmkTJMGH85TL0RpnwS6uaKIOVkC1Ojl4Rd6Cj3qGLtWvjHfxQxyE8p8iMfK1VbZvJkuO46uPlmuOQSiUxJpUREHSg8jPUIn7MN0ywvHgFSa2bXw7Dt+1IU2gxXJvKxFDMIzR+GaTdBy0eh7rxCEWpbpkUvtXdq+xRFGWZUGFIURakRfvc7+PrXpWaGP6tVJbnoIvjUp+C222REPJuVQqc+Gvkzisl1wrr/BTt/JI/92kCnMuX7mRAaD1Oug+m3wqSPAV6hiHOJ6miVGRUfbTw7acwVJfU8eP55+Ju/kZnCstnKzmRlGHDVVXDTTfDZz4oIbtv9BYSGhsqdj1JhUvvhvXth/7OAJcWJoXK2LzoFptwA59wOzUtluvlAfOyltY5B26cotcYYszqKoii1R3c3fP7z8LOfVWdGnA99CP75nyXNyK9JA1rrZ8yw+1F4988lIqfSM+IEErDgb2DuV6VGUbB+6OeMZsaYY7T/2FSW/wG8954Uk640114L3/++FIX2p6lXxgieB5vvh/X/W0Qgb6j8smEmMhEu+j9wznJJhw0WlMixIICXY4zZPkWpRVQYUhRFqSKOA5/+tMwmVo1peL/5Tbj7bnGKzFE+6YhShn0vwNv/T2XqBA2kbg584i2JDApoHaCxRjIT57/8v79mV1v/ukGV4rHHJEJy0NQiZXSz+dsiClVj9rDmD8M1P5Up1S0dgVEUpTZQYUhRFKWK/OY3Uli1GqLQnDlwzz1SOFUZo6z5SnVEIYAr/0OmildFckzyyMoVHOycUhVR6PrrVRQa03gurP0b8PLVOf7VT0i0pKIoSg2hd2OKoihVJB6vzmg5SNqa51Xn2EqNUG4mr0qRPVY9x6xWibRU+wwqRiKSxDSqY/w6O6W2kDJWMcCq4qwJuU5wK1hI62xgDNk+QV1wpfYwvBryCi677DJv9erV1T4NRRmbVPsuuYZsUaX54hfhkUeqU1/o1lvhRz+SqX51hp0xSMf78NrVxYKrlSQ0Dj76C0kpC+ro+VjDceDjH5c02krNtljK3XdLKm0wqLWFxiT7X4KVd1Qnlaxutti+0Hi1fYqiVBTDMNZ4nndZuW0qVyqKolSZ738f/uf/lOnDK+2gPPUUzJwJ//Ef4px1dVUvgkmpAuMuhmt/A/XzKx89lOuAVy6Ftz4nMwPlk+BUQSFQqoJlwcsvw5/+qaSzVjqj8P77Yf58eOklmQmtu7uyx1eqzLRPwx88J7OCWRWucdazHV6cBe/dLXYw3125WdAURVEGQSOGFEURNGKo6hw8CH/3d/Dgg/J19PRU9vixGFxzDSxbJgWx43Fx1rQG0RjAc2HXI7D5W5DcWZilp8KpDg0LYOqnYMZtsuxkIKTzhI8FtmyBv/1beO45sTmVnqGssVFmKLvlFrjuOrG/4TCEqphtpFQIJwvbvg9bvgO5Y2CngArejxgmjL+sYPtuhfhMcPMQrKvcOSiKMmY4UcSQCkOKoggqDNUM2Sy8+CL867/C734naV7pNNgVnk33ggukSOuHPwwXXQTTp8t5gBRt1ZrBo5TODbD9h7D/eci0gmFVPtUsWA+TPg4tH4MJV0D9XDAD4sRZUZ3JZ5TS0yOzhT34IKxeLeJMMlnZKEbDgEsugU9+Eq6+GhYtkunsUymxeXXqr49OPA/a3oLt/wIHfwp2L+BVvjh/uBmmXAcT/4sIRvVzRLh38xLVaWreo6Iop48KQ4qiDI0KQzWJbcM778BPfyqj6du3Sy2gVKryQlEgAHPniqN0ySVw1VUwbx40NIhg5DiSChevYj1j5cTYtlw7ritOdyAwRPpiaj8ceg32PQtHfw2YEklUjbockRZoXCSt+SoYdwnE79k58QAAGbhJREFUpkkRVzcr5xaIiYCk1CZ2SqLRDEsEvhN8V5kMrFwpqV4vvigRlaGQRBNVOt01EpG0s4UL4bLLYPFiOP98ibJMp+V8QiF5rNQm+Xyxjl84PERNPc+TdK9Dr8C+Z6DtbSlW7eTArUK6a2wGjCvYvqYlshyZJGKR/3sKxKRXag/PFdvn2WAEwIro/5RSNVQYUhRlaFQYOivo6IBVq2DdOunXroVDhyTdy/NkdL3ShEIwYwace6602bOlzZwJU6bA+PEiGuVy4kAZhogRkYhGHQ0X+bxEmvlioWXJ55vLQWsr7N8PO3fC1q3S79oFu3fDkSOn4GS7DnS+D+3vQfu70PYO9GwFPBnF9m98K01kIsTPhcS5EJ8lI+yJ80Q0ikwEMyxpaf65GQEwQ9WdlWg04bny+bp5JAXHkM/XDEKuHdKHoHc39GwTZ7t3NyR3Qe8euOPknexDh+DNN+H996XfsAHa2kSM8QXPShONip0791zp58wR23fOOTB5sqSo5XLy+/Q8sXe+7av2X+5oIZuVz9gpZL76Eymk03D0KOzdCzt2wLZtYvf81tZ2CgdxcnDsHehcKyJR+2pJuTUCYFpSH41KF+czpD5S4lxJP0vMgvrzpY9OhUiznJ+TKaQFGwVBVkWJYcNzCrbPBjxJCzRD0mePQepgwfZtheQOsXvJXZDaB8t1Rk6lOqgwpCgnw6RJ4iUpSqVpaYHDh0/76dms1OhYt07SL95+GzZvlhH3SERumCtds2Mgzc3F1tQkbeJEEY6mTpXlpiYYN66YqpHPy7m7blE3NAxxriyrGO1ytjtYjiPv1bblvZYKNYYh79WyRIDzi+S2t4vgc/gwHDggTnNbm6xra5N2+PDJfe9ndBvgeZA+AB3roHMdtL0pM52lD0pECEgqRjUEIx8rCtHJEG4qtGbpo5PFgYpOEicqNF5S2KxIYSTeBlwRPvow5abfCIjwYY6CEXonJ9+P5xTeq39BGIX3ahUcSVOK5OY6IdcGmaOQOiDfdeYoZNsKrVVappUha7XceWb3oL29sHGj2L5335X2wQfFCJ58vpj+Wg1MU2xbOds3darYv4kTYcIEsX2JhJx7Pl+0BYPZvtFQ/8i2i3becfrbItOU5r/XVEomR/Bt36FDYvsOHy7aPN/+HToktvJEnLH747ni9Hesg473xPZ1bpBr34oBXqFeURWnpQ/UiX0rZ/tiUyXqKNIsM0QG68EIgZeTQQDPod/v17cFvu0zzvJRHc8r1NLzbZ9HedsXlG12txQKz7ZB5kjR9mXbxNb5ti9zVATxoThD26cop8uICUOGYdwPfBrIATuA/+Z5Xmdh218DX0As4lc8z3t1qNdTYUipKme7d6mc3YyASN/aKk7Tzp2SgrZ5s4yaHjggtTyiUXEybLuYDlErxGLiJCUSkppWrveXGxulNTSIqFRXJ9tiMWnRqDgWpik/84HNNOXjH6z5pqHc86DovJU21xVHJ5OR1tsrjk1Pj7TubnFyOjtlOZmUfU7Ud3ePTPrgiIwPOVno3iJRIsmdstz9gThSmaOFcPqQ3Ie72UIqWI1gBMRJCsQLLVHsgwmw4sU+1ADBBgg2SrHYYH3J/rFCPaRIIcWjcAFilPSFi8gXZPqcE6/go3jFfRn4XGOQ5xV6Jyuj2U5K6qXYSWn57qLAk++CfE/J9oH9gHXDzQg4R54naWcbN0pkyPbtsGmTRI0cPCi2zp8FzReOamiMtM+2ncj2xeNi5xoaxPbV18uyv823fZGIpE2Vs3sjbfvSaWmpVH8bNtD2+dsGs329vbKfM8z6yoh953YvdG2S6LjkLujeJLYvtU+iSPxoOj/SzquhyBEzJHas1OYN1gfri4JSsKGwrq5g90ptn0lFbJ/nIVFbvu1Lg52W+nh2UuxcvkdsXq5D+oH2LZ+U/fMDbOBI1JlSYUipEiMpDH0C+KXnebZhGN8C8DzvXsMw5gOPAVcAU4DXgfM978RTnKgwpFQVFYaUalJhzySblRD7ffskzWjvXhGNdu6Ux0ePiqARDhejcvyR7MwonFHcd3YGNssqjtyXa6OBijvFnicjrr17Ib0feveJeJTcLulF6cNy424WUr4MC/AkgsfNVn62tErQ5zyZxy/7kTyeS18EU7/InrOYKjhHvb2wZ4/Yuf37ZXnrVhGRDhwQQd22izW4fNuXyw0dhXI2MlZtX1XEQNeRKJPUPqnfltoHPTsKtm+v2MV8V0E8KqQkeZ5EtTgZKp+uVgHU9ilKRTmRMHRGSaae571W8vAt4JbC8k3A457nZYFdhmFsR0SiN8/keIqiKMrwEA5LPYw5cwbfJ5eDY8f6pye1topodOCAjL4fPSr7dHZKJIxpSnROINB/9NhPlapVPK+YzjAa8Os4BQLi4PmRAX4tolBIogwaG6t0ctFJ0rii/D6eB/lOyJSmJxVC9jOHJIw/c1ge5zokzN/JyWxlRrA4ouw54lT11b+pUfocn2qfyDDRl25iFZw9Q74LNyd9IAGhalx8ElEzf760wUinT972dXVJdIuf8mRZRdtn28VWq4xG2+f/B/kilz+okc1KFFVDQ5VOzrQgPl3aYLhOwfa1Hm/70gcldTdzRNblOiQKxrOllpoZQH5rpbYvV7G3d1qMOtsXlO/BsEoGNfwJEihEllbH9inKUAxn9bHPA08UlqciQpHP/sI6RVEU5SwhFJICqpMnn9z+nifiUKkj5QtGyaQ4UB0dsq6jQ9ID/G29veKMZTJy8+7XD/JFjdKAvoFpC6VtYJ2KWqR0RN5fHpiq4eM7bX4tDhDHJhqV5qeUlKaVjB8vvb++rk5qmPg1TiZMEGGwpjEMSVMIjQNOoF6W4uQgd6x/zYe+FILuQgpBe0kKVXcxdcBOSbqAm5UP3QyKuNE3iu1TJnXBH8X2HGp/RN8oEWtKR+lL0jT68Irvy80XnU8rLKkigVgxpcRPJwmNF8En2FBMSQk1FmubRJql7kmNR+hGozBtmrSTwfPEvvlCUmurPC61fe3tYvt8m+inkKZSxbSrfL4o6PrCxsDjlLN9pcKOYbiEAjnCgSyhQO64FrTymIaLYXgYeNKXWe47JgaeV2iDLNtugJwdOq5l7TA5O4TjWoBxnN3zbZ//PksvC/+92bYMUpim2L5IROyen1bnC9yNjVKnaaDt8+s6NTeLbTzhLIy1gGlBeIK0k8VOD7B9xwppUD0ldcHaIddVtH1++pSTKqScZgt/QMFiTTFjoD0oZ/ucorhT05TYPsOimM5Wmqbm49s+WyJVPacwYUG0mCpXavtCjYU2vphOHKyT/y/f9oWbJA2vxm2fMrYZUhgyDON1YFKZTf/D87wXCvv8D8AGHvGfVmb/srfqhmF8EfgiwIwZM07ilBVlhGhp0eLTSnVoaan2GQwLhiE36fX1MGvWKTyxXOF3u9AqQFughU8sPHzCOhuDNSgu+zOBhcPFFo0WnRl/ORwW0S0U6r88sEWjxXpJvpMTCv3/7d19jKVnWcfx73Ve5m13urul7e6229raFgxYRLLWEBRRQIsiGCAGjaKRpDaBgPqH2uBrTIOJCfEPTQwR/yC8BQW1UrWWgCGNhaUtbaEv0mJbuixl2Xa3uzOzOy/nXP5xP2fm7Hb2pczOeWb3+X6Sp885zzkzc83Zp8+c+c19X7fvK0+qPVY1lD7DJBNOsejAfLWts20d+NjLOWmfoMEqX8cFNyf02hgca4+vBDetiSrAmahuV/0+OpMr01RO3Npj1aiD6n53cwlxhnuOnOsNZ9dJxEo4seoozMzyy/fyL+uD/ZGV8HLpCP2FGRZnD7I0d5DesUPkQukBFf15IheIXIRcoJWLBItELtJiiWCJVvQIegR9+rTJbNPPFv1skdmiX4U5yy1cADLK6bPKO/TBmff8b3bwYC7fLa9BCZVakaWW6NOiT0SPgFITbfrZIenQp0PGGBldkm7Zt8YgxsjOJmJsC63xrbSnttHdtI3O+NC52J2uzk3PUaD8f93ZVVZiPFMnvfaNaOjY1g58/Lqhax18X9e+aFXXrep6d9z1r7ruDbbB9a19smtg9XjnhHOtPeUPXjXCmlcli4jfAG4CXpeZc9WxmwEy8wPV/duBP8vMU04ls8eQJDXQRnjDtdGHGWl9eO7phcgsIy6Wp/kMTfU5OjTFcf5AGZWx6miMNtBeOfeWR5wNpjxu9JEXNYnu0PTEQVCw2qi2sSoImCpB0diFMLG9rMI1eWkZtTZ+MVzx9tN9xfOb1z6pkdatx1BE3AD8AfBTg1Cocivw8Yj4IKX59LXAnrV8LUmSJOmsyixBzqAZ8NzeqkHwYDnq/TBfTUFcmlmlKXrVPyRPM8QxqQKMUXxT56FchN4ZrODVXyjb4uHqwDdXf97Ednjr02etPEk61621x9DfAOPAHVGS5y9l5k2Z+WBEfAp4iDIZ4N2nW5FMkiRJOqv6SyXwmXkcZp8oy4gffrjcnttXeq9Eu0wjgRIq9E6x9OIgeNgIorMyhaq7GdrVfnhp8fbU6is+Dd8+7n77+Y1zswcMrRD1vFWiTrjdmz9+GfDV9qd6jUfhmK0DJGnYWlclu+YUj90C3LKWzy9JkiSd1uIMPPd1OPQAHNgDB+8tS4AvHoTWZBnp018qzcZPlEsrqwaNSmdzmdI0UTWmXW7QfQl0t8HYlqqR7fRK2NOeqvqmVD1VorUyhWq5CXBlOexpV42Ahvu0DN0f7tVyOrlKD6xVe8QMmhIPP1Y1/221y7SwaFUh3Hxp/N6bg1uvXvPLKkn6/pzNVckkSZKk9dPvwcw34dDX4OBX4cBdcOjrZeRPZ6oKf+aO/5je7Gh66ra6MHlZaQI8dXnZb7669LeZ2FFWmhrbWkIhsoQigwH10a4a6I69sK/ZHqcM3h+BGA6WzoJBU2DqWj9ekjRgMCRJkqSNaWkW9n8R9t4K3/08zD5eevzQLlOShpv2LPeVWUftSdjysrJNX7Oyn7ysjPDpHSsjeKJVnts6xfrorRcYAkmStE4MhiRJkrQxZMKh+2Hff8BTny4jg1oTZXn3wbSk/hk0IV6zgE1XwraXw9ZXwMWvgi3XlVWtlubKCJ/O1POnYLU2j6A2SZLOLoMhSZIk1evod+CxD8Ojf1tCoP7iSpPnUTV7ntwJO2+AK34ZLnnNylLonU2lR9HAmFOfJEnnF4MhSZIk1ePYAbj3d+Fb/wgE9Ee8WtXmH4Rrfhsuf1vpBdRfhO70aGvYKD6zw9W6JKmhDIYkSZI0evtuhzvfXs8S8GMXwo/cAle9s/QBGvQCak+Mto6NxFBIkhrLYEiSJEmjdfhRuPNtpbn0qHWm4YZ7YHJHs4MgSZIqrdM/RZIkSTqL9v17WVq+Dpe+EcYvMhSSJKliMCRJkqTRuvCVZWWvOjzzZZZXONOKie11VzBC/gokScOcSiZJkqTRuuQn4aU3w0MfgN7caL/27JPw+dfBqz4Ck7ug6xLzALz16borkCTVxLhckiRJo3fdH8Frb4NNV0JnxOHMM1+Bz74U9twIB++H3jwsHh5tDZIkbRCOGJIkSVI9tr8WfvFR+Pa/wSMfhGfuhgB6o1i2PuHJT5StuxV2vqEsW7/zBojqb6dNXbpektQokblx5ljv3r0777777rrLkCSNUkTdFcAG+lmoEfLc23hmn4K9/wLf+qfSC6g9DouzQG+ERQRsvQ62vQIu3A0X/Thc8EPQGiuBVXuy1CWdq7z2SY0UEfdk5u5VHzMYkiTVyjeoqovn3sbWW4AD/wPfvq2ERbNPQGcK+j3o1bDM/cQlsOU62PZyeNH1MP1imLocxraVwCiXSkPt9iS0HJRfu+xD7yj0F8sIsPZk+Xc6ug8ueEnd1dXLa5/USAZDkqSNyzeoqovn3rllaQ6eewgOPVCmnD27Bw4/UoKi9nh5PBdHX1d0YOqy0itpalcJi6avhc1XV8HRVuheAPRL2JVL1ce1yygkRx+duf4S9OfLniyBT6tbXselOVh8Do5+B2YehyOPlkbjR/fC7LdKsLhUBYq/2vD/77z2SY10qmDIP2dIUt127IDvfrfuKpptI7xJrkurBf1+3VU0V5PPve3b4ekXsBJWZwpetLtsV/9WOZYJx56Ggw+UwOjAXXDwPph7qoQFrS5krwoE1ukX0VwqAcTsk6epfzOMXwQTF5f9eLWf2F4CpYkd5bH2JuhMrkxZa42X0S+5VAKR7J3wvbSg1a6Cpm7ZbxSZZcROVnVnn5XaowQ7y3V3ynP782WkT+9oCXsWDpZ/47l9ZbTP/Pdg/kC1VbcXDlaf+wz5c7d+XvvqrkI6zpqCoYj4C+AtQB/YD/xmZu6rHrsZeBdlUvh7M/P2NdYqSecn35yqToZCqsvZuPZFwOTOsl36cyvHs19ChJknYPZxOPJNOPwwHPkGzO0tQcJgpElSjUKZX3s9p7I0U7bZJ174x7bGobu5hEbdzdDZVIKm1fbdC0oz7bEt5Vi0q2barZXbMXSb4WNDz6NfBVKDrQf0TjjWP+FYr7yOC8+V13jxuer7nh3az65ybKaEQOsV3p3In7uqk+efNqC1jhj6q8z8Y4CIeC/wJ8BNEfFS4B3Ay4BLgc9FxIszc5SdAyVJktRE0aqmde0CfuL5j2cfju0vI4vm9pb9YPrR0X0w/ywsHoTFmdIvqDVWhSVZwo/e/MqUsFHoz8P8PPDM6L6mJKkx1hQMZebhobubWIn53wJ8MjPngccj4jHgeuCutXw9SZIkac2iBZM7yvaiHzv58zJh8fDx05eOVbePPl361xx9ujy+8Gx5bm9+ZWpUtIemzGQJpPpLpReSfy9dg6he4+7QSCdYfo0H0+6gjKAa27YybW/yUuDv6ipckjakNfcYiohbgHcCzwE/XR2+DPjS0NP2VsckSZKkc0NEmZI1tgWmrzmzj+kvwuIRWDpSRhwtHanuz6zsl47AwqEy3WrhECweKqHS8uNz1Ypax0pzbfpVyNSmTPVqVYFTi9KrB6r/DMmqwW+u9PYZDqOWm//mCR8zOLba542V12XZ91vPidPUGGrIPVl6SnU2QWcautPQ3VIaeY9dWPbLx6eraXTT1TS74WNTJ/lHMhiSpGGnDYYi4nPAjlUeen9m/mtmvh94f9VT6D3An/L8nwRwkknDEXEjcCPAFVdccaZ1S9L5Y/t255urPjafVl22b6+7gvXR6sL4hWU7W/q9Msqov1BWNuuvsvXmVz9+3HMWKG/Jh0Kak96OKuyJU98eTLVbbWsP3x9f5djQdtzoqnXmz13V6Xy99umcdtpgKDNff4af6+PAbZRgaC9w+dBju4B9J/n8HwI+BGW5+jP8WpJ0/nBlCknSqbTaQBvaE9Ctu5jzgD93Jek4rdM/5eQi4tqhu28GHqlu3wq8IyLGI+Iq4Fpgz1q+liRJkiRJks6utfYY+suIeAllufongZsAMvPBiPgU8BCwBLzbFckkSZIkSZI2lrWuSva2Uzx2C3DLWj6/JEmSJEmS1s+appJJkiRJkiTp3GUwJEmSJEmS1FAGQ5IkSZIkSQ1lMCRJkiRJktRQBkOSJEmSJEkNZTAkSZIkSZLUUAZDkiRJkiRJDWUwJEmSJEmS1FAGQ5IkSZIkSQ1lMCRJkiRJktRQkZl117AsIr4HPFl3HWqsi4ADdRehRvLcU508/1QXzz3VxXNPdfL8U11+IDMvXu2BDRUMSXWKiLszc3fddah5PPdUJ88/1cVzT3Xx3FOdPP+0ETmVTJIkSZIkqaEMhiRJkiRJkhrKYEha8aG6C1Bjee6pTp5/qovnnuriuac6ef5pw7HHkCRJkiRJUkM5YkiSJEmSJKmhDIbUeBFxQ0T8b0Q8FhF/WHc9ao6I+IeI2B8RX6+7FjVLRFweEV+IiIcj4sGIeF/dNak5ImIiIvZExP3V+ffnddekZomIdkR8NSI+W3ctapaIeCIivhYR90XE3XXXIw04lUyNFhFt4BvAG4C9wFeAX8nMh2otTI0QEa8BZoCPZOYP112PmiMidgI7M/PeiJgG7gF+yWufRiEiAtiUmTMR0QXuBN6XmV+quTQ1RET8HrAbuCAz31R3PWqOiHgC2J2ZB+quRRrmiCE13fXAY5n5f5m5AHwSeEvNNakhMvOLwLN116HmyczvZOa91e0jwMPAZfVWpabIYqa62602/1KpkYiIXcAvAH9fdy2StFEYDKnpLgOeGrq/F385ktQgEXEl8KPAl+utRE1STeW5D9gP3JGZnn8alb8Gfh/o112IGimB/4qIeyLixrqLkQYMhtR0scox/2opqREiYjPwaeB3MvNw3fWoOTKzl5mvAHYB10eE02m17iLiTcD+zLyn7lrUWK/OzFcCbwTeXbUVkGpnMKSm2wtcPnR/F7CvplokaWSq3i6fBj6WmZ+pux41U2YeAv4buKHmUtQMrwbeXPV5+STwMxHx0XpLUpNk5r5qvx/4Z0pbC6l2BkNquq8A10bEVRExBrwDuLXmmiRpXVXNfz8MPJyZH6y7HjVLRFwcEVur25PA64FH6q1KTZCZN2fmrsy8kvKe7/OZ+Ws1l6WGiIhN1YIPRMQm4GcBV6bVhmAwpEbLzCXgPcDtlOarn8rMB+utSk0REZ8A7gJeEhF7I+Jdddekxng18OuUv5bfV20/X3dRaoydwBci4gHKH2juyEyXDZd0vtsO3BkR9wN7gNsy8z9rrkkCXK5ekiRJkiSpsRwxJEmSJEmS1FAGQ5IkSZIkSQ1lMCRJkiRJktRQBkOSJEmSJEkNZTAkSZIkSZLUUAZDkiRJkiRJDWUwJEmSJEmS1FAGQ5IkSZIkSQ31/2bY47qOUqZsAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1440x144 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#plot the PSSM \n",
    "pssm=np.asarray([pssm['A'],pssm['C'],pssm['G'],pssm['T']])\n",
    "print(pssm)\n",
    "plot_weights(pssm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The inverse problem: Scanning a DNA sequence with a PSSM to identify good matches i.e. likely binding instances<a name='Scan' />\n",
    "\n",
    "The PSSM encodes the log odds of the probability of observing a specific nucleotide at each position across a collection of subsequences that a TF would bind to, relative to genomic background. \n",
    "\n",
    "Now, lets invert the problem.\n",
    "\n",
    "#### We are given a PSSM of a TF (of length 'k') and we are given a long sequence (say an entire chromosome). We want to find all the positions in the long sequence that are strong matches to the PSSM i.e. positions to which the TF is likely to bind.\n",
    "\n",
    "Formally, we can rephrase this statement as a concrete set of steps (an algorithm):\n",
    "\n",
    "We are given a PWM of a TF (of length 'k') and we are given a long sequence (say an entire chromosome) (of length 'N'). \n",
    "1. We scan this DNA sequence, one position at a time (i = 1 to N)\n",
    "2. At each position 'i', we obtain a subsequence of length 'k', starting at 'i' and ending at 'i+k-1'\n",
    "3. We score this subsequence in terms of how good of a match it is to the PSSM\n",
    "4. We decide some threshold on this score to decide if it is a good match to the PSSM or not\n",
    "5. Positions that pass the threshold are good matches (likely binding sites) of the TF.\n",
    "\n",
    "Steps (1) and (2) are obvious. \n",
    "\n",
    "### Computing the match score of a subsequence \n",
    "\n",
    "How do we perform Step (3) i.e. how do we obtain a match score for a sequence of length 'k' to a PSSM of length 'k'.\n",
    "\n",
    "We simply sum the log odds scores of each of the nucleotides observed in the sequence at each position.\n",
    "\n",
    "Lets say our sequence was CGATA and we wanted to score this sequence relative to the PSSM of length 5 we obtained above. \n",
    "Position 1 in the sequence is a C. Looking up the log odds of C (row 2) in position 1 (col 1) of the PSSM we get a score of 1\n",
    "\n",
    "Position 2 in the sequence is a G. Looking up the log odds of G (row 3) in position 2 (col 2) of the PSSM we get a score of 2\n",
    "\n",
    "Position 3 in the sequence is a A. Looking up the log odds of A (row 1) in position 3 (col 3) of the PSSM we get a score of 2\n",
    "\n",
    "Position 4 in the sequence is a T. Looking up the log odds of A (row 1) in position 4 (col 4) of the PSSM we get a score of 2\n",
    "\n",
    "Position 5 in the sequence is a A. Looking up the log odds of A (row 1) in position 5 (col 5) of the PSSM we get a score of 2\n",
    "\n",
    "Match score = 1+2+2+2+2 = 9\n",
    "\n",
    "This is a high positive score! Which means CGATA has a greater odds of matching the PWM than the background of the genome.\n",
    "\n",
    "So CGATA must be a good match to the PSSM. If you visually cross check with the PWM or PSSM logo you will notice that CGATA is infact a good match.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Question: \n",
    "\n",
    "Calculate the score for CGGGG. Is it a good match to the PSSM?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Motif scanning with Biopython <a name='Biopython' />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Going back to the five steps above, how can we write a program to scan a longer sequence with the PSSM to obtain a row of match scores for each position in the sequence? We have outlined the way an algorithm could work in the Appendix at the end of this class, but we will not be covering these steps in class. Biopython, however has a built-in functionality for motif scanning. To perform the motif scanning, we can simply run the following commands in Biopython: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Position 3: score = -14.937\n",
      "Position -12: score = -49.836\n",
      "Position 4: score = -14.937\n",
      "Position -11: score = -25.904\n",
      "Position 5: score = -49.836\n",
      "Position -10: score = -26.903\n",
      "Position 6: score = -37.870\n",
      "Position -9: score = -14.937\n",
      "Position 7: score = -49.836\n",
      "Position -8: score = -2.971\n",
      "Position 8: score = -37.870\n",
      "Position -7: score = -49.836\n",
      "Position 9: score = -38.870\n",
      "Position -6: score = -49.836\n",
      "Position 10: score = 8.995\n",
      "Position -5: score = -49.836\n",
      "Position 11: score = -26.903\n",
      "Position -4: score = -25.904\n"
     ]
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "sequence=\"NNGCATTACCGATAA\" ## We pad the sequence with NN so that the motif can be centered at position 0 in the \n",
    "                           ## original sequence without sliding off the edge of the sequence. i.e. the PWM is scanned \n",
    "                           ## against NNGCA. \n",
    "for position, score in m.pssm.search(sequence, threshold=-100):\n",
    "    print(\"Position %d: score = %5.3f\" % (position+1, score))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The negative positions refer to instances of the motif found on the reverse strand of the test sequence, and follow the Python convention on negative indices. You can also calculate the scores at all positions along the sequence: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([       nan,        nan, -14.937336, -14.937336, -49.836132,\n",
       "       -37.869987, -49.836132, -37.869987, -38.869625,   8.994954,\n",
       "       -26.90348 ], dtype=float32)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m.pssm.calculate(sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In general, this is the fastest way to calculate PSSM scores. The scores returned by pssm.calculate are for the forward strand only. To obtain the scores on the reverse strand, you can take the reverse complement of the PSSM: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([        nan,         nan, -49.836132 , -25.903841 , -26.90348  ,\n",
       "       -14.937336 ,  -2.9711912, -49.836132 , -49.836132 , -49.836132 ,\n",
       "       -25.903841 ], dtype=float32)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpssm=m.pssm.reverse_complement()\n",
    "rpssm.calculate(sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How do we map genome-wide transcription factor binding using ChIP-seq experiments ? <a name='ChIPseq' />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ChIP-sequencing, also known as ChIP-seq, is a method used to analyze protein interactions with DNA. ChIP-seq combines a technique known as chromatin immunoprecipitation (ChIP) with massively parallel DNA sequencing to identify the binding sites of DNA-associated proteins. It can be used to map global binding sites precisely for any protein of interest.\n",
    "\n",
    "ChIP is a powerful method to selectively enrich for DNA sequences bound by a particular protein in living cells. First, formaldehyde is used to cross-link (covalent bond) all proteins bound to DNA i.e. the proteins are not stuck to the DNA. The DNA is then sheared into tiny fragments, some of which are bound to the protein and most are not. An antibody targeting the protein is then used to pull down (enrich for) the fragments bound by the protein. This is called <b>chromatin immunoprecipitation (or ChIP)</b>. All the other DNA fragments are washed away. The protein is then unlinked from the bound DNA fragments and washed away.\n",
    "\n",
    "Hence, the DNA obtained from ChIP is mainly enriched for bound DNA fragments. All the resulting ChIP-DNA fragments are sequenced simultaneously using a high-throughput sequencer.\n",
    "\n",
    "The sequencing reads are then mapped to a reference genome using a short-read mapper such as Bowtie.\n",
    "\n",
    "Regions with a large number of reads mapping to them are likely binding sites of the protein.\n",
    "\n",
    "For any ChIP-seq experiment, a corresponding <b>control experiment</b> is performed, where the exact same procedure is followed as above except the antibody is not used to pull down the protein. So the DNA fragments in the control experiment are just sampled non-specifically from the genome. But this experiment allows us to observe any biases that may occur in terms of sampling reads at random. In fact, it has been observed that certain regions of the genome tend to be more easily sheared and sequenced than others. It is important to correct the ChIP-seq signal with the control to avoid false positives binding site positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<iframe src=\"https://drive.google.com/file/d/0B_ssVVyXv8ZSSkhvVUlYc01zOGc/preview\" width=\"1000\" height=\"480\"></iframe>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import HTML\n",
    "HTML('<iframe src=\"https://drive.google.com/file/d/0B_ssVVyXv8ZSSkhvVUlYc01zOGc/preview\" width=\"1000\" height=\"480\"></iframe>')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "We obtain four types of outputs from a ChIP-seq experiment\n",
    "\n",
    "* <b>Genome-wide count coverage tracks for ChIP and control</b>: which counts the number of ChIP-fragments and control experiment fragments overlapping each position in the genome, respectively. \n",
    "\n",
    "* <b>A peaks track with statistically significant regions of signal enrichment</b>: To identify high confidence binding sites, the coverage track is then analyzed using a statistical model that compares the observed ChIP-seq read counts at each position to the control read counts at the same position (adjusted for differences in total number of reads).\n",
    "\n",
    "* <b>Genome-wide normalized fold-enrichment coverage track</b>: which is similar to the count coverage track. But instead of raw counts at each position, it has the fold-enrichment of ChIP-seq counts relative to control counts\n",
    "\n",
    "\n",
    "* <b>Genome-wide -log10(p-value) coverage track</b>: which is similar to the count coverage track. But instead of raw counts at each position, it has the -log10(p-value) of statistical significance of ChIP-seq counts relative to control counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "<b>Question:</b> If I had 10,000 sequencing reads from a ChIP-seq experiment and a genome of size 1000 bp, how many reads should I expect by chance if I uniformly scattered the reads along the genome?\n",
    "\n",
    "The ChIP-seq signal at position in the genome is computed as the fold-enrichment i.e. ratio of observed ChIP-seq counts to control counts. This is also referred to as the *effect-size* of signal.\n",
    "\n",
    "Further, we use a statistical null model (based on a count based random distribution called the Poisson distribution) to compute a *p-value* for the observed count at each position i.e. what is the probability that the counts >= the observed value could have just happened by chance.\n",
    "\n",
    "We then use multiple hypothesis testing correction using the Benjamini Hochberg method to obtain adjusted *p*-values. We use a threshold e.g. 0.05 on the adjusted p-values to identify genomic locations with statistically significant signal. These are the likely binding sites of the transcription factor!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Visualizing ChIP-seq data in the genome browser <a name='ChipseqBrowser' />"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets visualize ChIP-seq data for a transcription factor called CTCF in a lymphoblastoid cell-line called GM12878.\n",
    "\n",
    "WashU Epigenome browser session: http://epigenomegateway.wustl.edu/browser/?genome=hg19&session=ddNxKsyNNX&statusId=1299889156"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## APPENDIX"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### One-hot Encoding\n",
    "\n",
    "Going back to the five steps above, how can we write a program to scan a longer sequence with the PSSM to obtain a row of match scores for each position in the sequence?\n",
    "\n",
    "To do this efficiently, we will use a slightly different representation for the DNA sequence called a 'one-hot-encoding'.\n",
    "\n",
    "<b>'One-hot-encoding'</b> is a process that transforms a sequence of nucleotides of length 'N' into a matrix of 1's and 0's with 4 rows (representing A,C,G and T) and 'N' columns (representing positions).\n",
    "\n",
    "In the graphic below, a colored square represents a '1', indicating the presence of the specified nucleotide at that position along the sequence. A blank square represents a '0', indicating the absence of the specified base at the position. \n",
    "![one-hot-encoding](images/10_one_hot.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1],\n",
       "       [0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0],\n",
       "       [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0],\n",
       "       [0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0]])"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#we write a function to perform one-hot encoding. \n",
    "def one_hot_encode(sequence): \n",
    "    encoding_dict=dict() \n",
    "    encoding_dict['A']=[1,0,0,0]\n",
    "    encoding_dict['C']=[0,1,0,0]\n",
    "    encoding_dict['G']=[0,0,1,0]\n",
    "    encoding_dict['T']=[0,0,0,1]\n",
    "    encoding_dict['N']=[0,0,0,0]\n",
    "    encoded_sequence=[] \n",
    "    for base in sequence: \n",
    "        encoded_sequence.append(encoding_dict[base])\n",
    "    return np.array(encoded_sequence).transpose()\n",
    "\n",
    "#We now one-hot encode a DNA sequence \n",
    "sequence=\"NNGCATTACCGATAA\"\n",
    "encoded_sequence=one_hot_encode(sequence)\n",
    "encoded_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The inner product\n",
    "We now scan the PSSM (length 5) across the entire sequence by aligning the PSSM with every subsequence of length 5 in the sequence starting at position 0 and computing the match score as we did above.\n",
    "\n",
    "Note that the match score for a subsequence is simply the sum of position-wise scores from the PSSM corresponding to each nucleotide in the subsequence.\n",
    "\n",
    "An efficient command to obtain this match score of a subsequence aligned to a PSSM is an <b>inner product</b> of the PSSM matrix with the 1-hot encoded matrix of the aligned subsequence. \n",
    "\n",
    "An inner product is simply a sum of the element-by-element multiplication of the values in the two matrices i.e. the PSSM matrix (4 rows, 5 columns) and the one-hot encoded subsequence (4 rows, 5 columns).\n",
    "\n",
    "An inner product is often represented mathematically as a < , > operator. E.g. if 'W' is the PSSM matrix and 'X' is the one-hot encoded subsequence, their inner-product is written as < W , X >\n",
    "\n",
    "![Convolution product](../Images/10_convolution3.png)\n",
    "\n",
    "#### Question:\n",
    "Can you explain how the inner product of the PSSM and the one-hot encoded sequence is equivalent to the sum of position-wise scores from the PSSM corresponding to each nucleotide in the subsequence.\n",
    "\n",
    "\n",
    "### Convolution operation\n",
    "\n",
    "There is also a built in operation known as the <b>'convolution'</b> that can perform a scanning inner-product of one short matrix (PSSM) against a longer matrix (one-hot encoded longer sequence). The convolution directly outputs the match scores at each position in the sequence.\n",
    "\n",
    "We can then simply restrict to positions with positive match scores (log-odds) as the positions that will be most likely bound by the TF.\n",
    "\n",
    "\n",
    "![Convolution formula](../Images/10_optimal_convolution_product.png)\n",
    "\n",
    "\n",
    "We can perfom the motif scan in Python by following the steps below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false,
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Convolution product:[-10. -20. -18. -28. -15. -15. -50. -38. -50. -38. -39.   9. -27. -28.\n",
      " -18.]\n",
      "Motif starting position along the sequence:9\n"
     ]
    }
   ],
   "source": [
    "from scipy.signal import fftconvolve\n",
    "import math\n",
    "def scan_sequence(sequence,pssm): \n",
    "    convolution_product=fftconvolve(sequence[::-1,::-1],pssm,mode=\"same\")[::-1,::-1][1]\n",
    "    #get the starting position of the motif along the sequence \n",
    "    starting_pos=np.argmax(convolution_product)-math.floor(pssm.shape[1]/2)\n",
    "    return convolution_product,starting_pos\n",
    "\n",
    "convolution_product,starting_pos=scan_sequence(encoded_sequence,pssm)\n",
    "print(\"Convolution product:\"+str(convolution_product))\n",
    "print(\"Motif starting position along the sequence:\"+str(starting_pos))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The best alignment is achieved by aligning the first base of the PWM to position 9 (0-indexed) in NNGCATTACCGATAA, \n",
    "at subsequence \"CGATA\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
